<!-- TOC -->

- [å•ç»†èƒæ•°æ®åˆ†æ](#å•ç»†èƒæ•°æ®åˆ†æ)
  - [ä¸Šæ¸¸åˆ†æ](#ä¸Šæ¸¸åˆ†æ)
    - [æ•°æ®ä¸‹è½½](#æ•°æ®ä¸‹è½½)
      - [è½¯ä»¶ç¯å¢ƒ](#è½¯ä»¶ç¯å¢ƒ)
      - [ä½¿ç”¨sratoolsä¸­çš„prefetch](#ä½¿ç”¨sratoolsä¸­çš„prefetch)
      - [åˆ©ç”¨ascpç”±ftp.ncbiä¸‹è½½æµ‹åºæ•°æ®](#åˆ©ç”¨ascpç”±ftpncbiä¸‹è½½æµ‹åºæ•°æ®)
      - [æå–fastqæ–‡ä»¶](#æå–fastqæ–‡ä»¶)
    - [Cell Ranger](#cell-ranger)
      - [Cell Rangerè½¯ä»¶å®‰è£…](#cell-rangerè½¯ä»¶å®‰è£…)
      - [å‚è€ƒåŸºå› ç»„åºåˆ—çš„ä¸‹è½½](#å‚è€ƒåŸºå› ç»„åºåˆ—çš„ä¸‹è½½)
      - [mkfastq æ‹†åˆ†æ•°æ®](#mkfastq-æ‹†åˆ†æ•°æ®)
      - [ğŸ”ºcount ç»†èƒå’ŒåŸºå› çš„å®šé‡](#count-ç»†èƒå’ŒåŸºå› çš„å®šé‡)
      - [aggrå¤šä¸ªæ–‡åº“çš„æ•´åˆ](#aggrå¤šä¸ªæ–‡åº“çš„æ•´åˆ)
    - [å…³äºä¸Šæ¸¸åˆ†æå¾—åˆ°çš„æ•°æ®é›†](#å…³äºä¸Šæ¸¸åˆ†æå¾—åˆ°çš„æ•°æ®é›†)
  - [æ•°æ®é¢„å¤„ç†å·¥ä½œæµç¨‹](#æ•°æ®é¢„å¤„ç†å·¥ä½œæµç¨‹)
    - [QCè´¨é‡æ£€æµ‹](#qcè´¨é‡æ£€æµ‹)
    - [å½’ä¸€åŒ–æ•°æ®](#å½’ä¸€åŒ–æ•°æ®)
    - [è¯†åˆ«é«˜å¯å˜åŸºå› ](#è¯†åˆ«é«˜å¯å˜åŸºå› )
    - [æ ‡å‡†åŒ–æ•°æ®](#æ ‡å‡†åŒ–æ•°æ®)
    - [æ‰§è¡Œçº¿æ€§é™ç»´](#æ‰§è¡Œçº¿æ€§é™ç»´)
    - [ç¡®å®šæ•°æ®é›†çš„â€œç»´åº¦â€](#ç¡®å®šæ•°æ®é›†çš„ç»´åº¦)
  - [æ•´åˆå¤šä¸ªscRNA-seqæ•°æ®é›†](#æ•´åˆå¤šä¸ªscrna-seqæ•°æ®é›†)
      - [ä¸seuratæ ‡å‡†å¤„ç†æµç¨‹ä¸€è‡´](#ä¸seuratæ ‡å‡†å¤„ç†æµç¨‹ä¸€è‡´)
      - [ä½¿ç”¨ SCTransform è§„èŒƒåŒ–çš„æ•°æ®é›†æ‰§è¡Œæ•´åˆ](#ä½¿ç”¨-sctransform-è§„èŒƒåŒ–çš„æ•°æ®é›†æ‰§è¡Œæ•´åˆ)
  - [å°† Seurat ä¸å¤šæ¨¡å¼æ•°æ®ä¸€èµ·ä½¿ç”¨](#å°†-seurat-ä¸å¤šæ¨¡å¼æ•°æ®ä¸€èµ·ä½¿ç”¨)
  - [ä¸‹æ¸¸åˆ†æ](#ä¸‹æ¸¸åˆ†æ)
    - [ç»†èƒæ°´å¹³](#ç»†èƒæ°´å¹³)
      - [ç°‡](#ç°‡)
        - [èšç±»åˆ†æ](#èšç±»åˆ†æ)
        - [å¯¹ç°‡è¿›è¡Œç»†èƒç±»å‹æ³¨é‡Š](#å¯¹ç°‡è¿›è¡Œç»†èƒç±»å‹æ³¨é‡Š)
      - [æ„å»ºå•ç»†èƒè½¨è¿¹](#æ„å»ºå•ç»†èƒè½¨è¿¹)
    - [åŸºå› æ°´å¹³](#åŸºå› æ°´å¹³)
      - [åŸºå› çš„å·®å¼‚è¡¨è¾¾åˆ†æ](#åŸºå› çš„å·®å¼‚è¡¨è¾¾åˆ†æ)
        - [å¯»æ‰¾å·®å¼‚è¡¨è¾¾çš„ç‰¹å¾åŸºå› ï¼ˆç°‡ç”Ÿç‰©æ ‡å¿—ç‰©ï¼‰](#å¯»æ‰¾å·®å¼‚è¡¨è¾¾çš„ç‰¹å¾åŸºå› ç°‡ç”Ÿç‰©æ ‡å¿—ç‰©)
        - [ä¸ä¼ªæ—¶é—´åˆ†æç›¸å…³çš„å·®å¼‚è¡¨è¾¾åˆ†æ](#ä¸ä¼ªæ—¶é—´åˆ†æç›¸å…³çš„å·®å¼‚è¡¨è¾¾åˆ†æ)
      - [åŸºå› é›†åˆ†æ](#åŸºå› é›†åˆ†æ)
        - [è¡¨è§‚é—ä¼ åˆ†æ](#è¡¨è§‚é—ä¼ åˆ†æ)
        - [å¯Œé›†åˆ†æ](#å¯Œé›†åˆ†æ)
      - [åŸºå› è°ƒæ§ç½‘ç»œåˆ†æ](#åŸºå› è°ƒæ§ç½‘ç»œåˆ†æ)
        - [åŸºå› å…±è¡¨è¾¾](#åŸºå› å…±è¡¨è¾¾)
        - [è›‹ç™½è´¨äº’ä½œç½‘ç»œ](#è›‹ç™½è´¨äº’ä½œç½‘ç»œ)
        - [å›å½’æ¨¡å‹](#å›å½’æ¨¡å‹)
    - [](#)

<!-- /TOC -->
# å•ç»†èƒæ•°æ®åˆ†æ
![](images/2021-10-22-21-29-10.png)
<br>ä¸€ä¸ªæ²¹æ»´ (GEM)=ä¸€ä¸ªå•ç»†èƒ+ä¸€ä¸ªå‡èƒ¶å¾®ç =ä¸€ä¸ªscRNA-Seqï¼Œå¯ä»¥è¯´è¿™å°±æ˜¯10Xçš„åŸºæœ¬æŠ€æœ¯åŸç†ã€‚

![](images/2021-10-22-21-37-28.png)

<br> Gel beadsæ˜¯ç”±å‡èƒ¶ç£ç å’Œç£ç ä¸Šçš„ä¸€æ®µå¼•ç‰©æ„æˆï¼Œå¼•ç‰©åºåˆ—æ„æˆä¾æ¬¡ä¸ºï¼š
* å…¨é•¿Illumina TruSeq Read 1 æµ‹åºå¼•ç‰©ï¼ˆR1ï¼‰
* 16nt 10X Barcodeåºåˆ—(Barcode)ï¼ˆæ¯ä¸ªGel beadçš„10X Barcodeå‡ä¸ç›¸åŒï¼Œå½¢æˆGEMï¼ˆGel Beads-in-emulsionï¼‰åç”¨äºåŒºåˆ†ç»†èƒï¼‰
* 12 nt unique molecular identifier (UMI) ï¼ˆåŒºåˆ†åŒä¸€ç»†èƒçš„ä¸åŒè½¬å½•æœ¬å¹¶å»é™¤PCR Duplicationsï¼Œå®ç°ç»å¯¹å®šé‡ï¼‰
* 30 nt poly dTåè½¬å½•å¼•ç‰©
## ä¸Šæ¸¸åˆ†æ
æœ€å¸¸ç”¨ä¸”æœ€å¸¸è§çš„çš„æ•°æ®é›†ä¸»è¦æ¥è‡ª10X Genomicså’ŒSmart-seq2ï¼Œç›®å‰å¤§éƒ¨åˆ†çš„å•ç»†èƒæ•°æ®éƒ½æ¥è‡ª10X Genomicsï¼Œå› æ­¤åç»­çš„æ•°æ®å¤„ç†å’Œåˆ†æä»¥è¿™è¯¥ç±»æ•°æ®ä¸ºä¾‹å­è¿›è¡Œè®²è§£ã€‚
### æ•°æ®ä¸‹è½½
#### è½¯ä»¶ç¯å¢ƒ
åŸå§‹æ•°æ®ä¸€èˆ¬æ˜¯ä»¥SRRæ ¼å¼å­˜æ”¾ï¼Œè¿™ä¸ªæ–‡ä»¶ä¸€èˆ¬éƒ½è¦å‡ ä¸ªGï¼Œäºæ˜¯ä¸‹è½½å™¨é¦–é€‰ascpï¼Œä½†æ˜¯ç›´æ¥ä½¿ç”¨ascpä¸‹è½½åˆéœ€è¦é…ç½®ä¸€äº›å‚æ•°ï¼Œå¯¹äºæ–°æ‰‹æ¥è¯´ï¼Œæœ€å¥½æ˜¯èƒ½æä¾›ä¸€ä¸ªIDï¼Œç„¶åç›´æ¥å°±ä¸‹è½½ï¼Œè¿™ä¸ªå°±éœ€è¦ç”¨åˆ°`prefetch` ä¸`ascp`çš„ç»„åˆäº†ã€‚
1. prefetch
   é»˜è®¤æƒ…å†µä¸‹ï¼Œ`prefetch`æ˜¯åˆ©ç”¨httpsæ–¹å¼å»ä¸‹è½½åŸå§‹æ•°æ®ã€‚
    ```bash
    conda install -c daler sratoolkit
    prefetch -h # å¯ä»¥æ˜¾ç¤ºå¸®åŠ©æ–‡æ¡£å°±è¯´æ˜å®‰è£…æˆåŠŸ
    # å¦‚æœè¦ä¸‹è½½æ•°æ®æ¯”å¦‚SRRæ–‡ä»¶ï¼Œç›´æ¥åŠ IDå·ï¼ŒæŒ‡å®šè¾“å‡ºç›®å½•å°±å¥½
    prefetch SRRxxxxxxx -O PATH
    ```
2. ascp
`prefetch`é€Ÿåº¦æœ‰ä¸€å®šçš„é™åˆ¶ã€‚å› æ­¤æˆ‘ä»¬éœ€è¦å…ˆå®‰è£…ä¸€æ¬¾å«åš"aspera"çš„ä¸‹è½½å·¥å…·ï¼Œå®ƒæ˜¯IBMæ——ä¸‹çš„å•†ä¸šé«˜é€Ÿæ–‡ä»¶ä¼ è¾“è½¯ä»¶ï¼Œä¸NCBIå’ŒEBIæœ‰åä½œåˆåŒã€‚
    ```bash
    wget http://download.asperasoft.com/download/sw/connect/3.7.4/aspera-connect-3.7.4.147727-linux-64.tar.gz
    tar zxvf aspera-connect-3.7.4.147727-linux-64.tar.gz
    #å®‰è£…
    bash aspera-connect-3.7.4.147727-linux-64.sh
    # ç„¶åcdåˆ°æ ¹ç›®å½•ä¸‹çœ‹çœ‹æ˜¯ä¸æ˜¯å­˜åœ¨äº†.asperaæ–‡ä»¶å¤¹ï¼Œæœ‰çš„è¯è¡¨ç¤ºå®‰è£…æˆåŠŸ
    cd && ls -a
    # å°†asperaè½¯ä»¶åŠ å…¥ç¯å¢ƒå˜é‡ï¼Œå¹¶æ¿€æ´»
    echo 'export PATH=~/.aspera/connect/bin:$PATH' >> ~/.bashrc
    source ~/.bashrc
    # æœ€åæ£€æŸ¥ascpæ˜¯ä¸æ˜¯èƒ½ç”¨äº†
    ascp --help
    ```
å¦‚æœæŠ¥é”™ï¼š
```bash
ascp: Failed to open TCP connection for SSH, exiting.

Session Stop  (Error: Failed to open TCP connection for SSH)

# å®˜ç½‘ç»™å‡ºçš„è§£å†³åŠæ³•æ˜¯ï¼šhttps://support.asperasoft.com/hc/en-us/articles/216126918-Error-44-UDP-session-initiation-fatal-error
On many Linux systems the default firewall can be configured with iptables. You will have to allow all incoming and outgoing traffic on UDP port 33001 (or whatever your Aspera UDP port is), which you can do with the following commands:
# ä½¿ç”¨ä¸‹é¢è¿™ä¸¤ä¸ªå‘½ä»¤(ä½†éœ€è¦ç®¡ç†å‘˜æƒé™)
# iptables -I INPUT -p tcp --dport 33001 -j ACCEPT
# iptables -I OUTPUT -p tcp --dport 33001 -j ACCEPT
```
#### ä½¿ç”¨sratoolsä¸­çš„prefetch
ä»¥GSE117988çš„æ•°æ®é›†ä¸ºä¾‹ã€‚
1. æ‰“å¼€https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE117988
2. ç‚¹å‡»SRAè¿™é‡Œçš„`SRP155988`
![](images/2021-10-22-21-04-05.png)
3. `send to` => `Run Selector` => `Go` 
![](images/2021-10-22-21-04-47.png)
4. ä¸‹è½½Accession Listï¼Œç„¶åå°±å¾—åˆ°äº†ä¸€ä¸ªæ–‡æœ¬æ–‡ä»¶ï¼Œåˆ—å‡ºäº†6ä¸ªSRR IDå·
![](images/2021-10-22-21-05-12.png)
5. ä¸‹è½½ä»£ç 
```bash
wkd=/home/project/single-cell/MCC
cd $wkd/raw
# for patient 2586-4
cat >SRR_Acc_List-2586-4.txt
SRR7722937
SRR7722938
SRR7722939
SRR7722940
SRR7722941
SRR7722942

cat SRR_Acc_List-2586-4.txt |while read i
do prefetch $i -O `pwd` && echo "** ${i}.sra done **"
done
# ä¸€èˆ¬2.6Gæ–‡ä»¶ä¸‹è½½2åˆ†é’Ÿå·¦å³
```
#### åˆ©ç”¨ascpç”±ftp.ncbiä¸‹è½½æµ‹åºæ•°æ®
1. è¿›å…¥å®˜ç½‘https://www.ebi.ac.uk/ena ï¼Œæœç´¢æƒ³ä¸‹è½½çš„SRAå·
![](images/2021-10-22-21-07-38.png)
2. é€‰æ‹©SRRè¿™é‡Œ
![](images/2021-10-22-21-07-56.png)
1. EBIå¯ä»¥ç›´æ¥ä¸‹è½½fastqæ ¼å¼æ–‡ä»¶ï¼ˆå·¦è¾¹æ–¹æ¡†ï¼‰ï¼Œå¦‚æœè¦ä¸‹è½½sraå°±å¤åˆ¶å³è¾¹çº¢è‰²æ–¹æ¡†ä¸­é“¾æ¥
![](images/2021-10-22-21-08-39.png)
4. ç„¶ååˆ©ç”¨è¿™ä¸ªä»£ç ä¸‹è½½
```bash
ascp -QT -l 300m -P33001 -i ~/.aspera/connect/etc/asperaweb_id_dsa.openssh era-fasp@fasp.sra.ebi.ac.uk:vol1/srr/SRR772/009/SRR7722939 ./
```
#### æå–fastqæ–‡ä»¶
æˆ‘ä»¬ä½¿ç”¨fastq-dumpè¿™æ¬¾è½¯ä»¶ï¼Œå®ƒæ˜¯sra-toolä¸­çš„ä¸€ä¸ªå·¥å…·ï¼Œä½¿ç”¨condaå®‰è£…å³å¯ã€‚
```bash
conda install -c bioconda sra-tools
```
```bash
wkd=/home/project/single-cell/MCC ///è·¯å¾„
cd $wkd/raw/P2586-4
cat SRR_Acc_List-2586-4.txt |while read i
do
time fastq-dump --gzip --split-files -A $i ${i}.sra && echo "** ${i}.sra to fastq done **"
done
```
* --gzipå°†ç”Ÿæˆçš„ç»“æœfastqæ–‡ä»¶è¿›è¡Œå‹ç¼©
* --split-filesï¼šé¦–å…ˆå®ƒæ˜¯åˆ†å‰²çš„æ„æ€ï¼Œ-3å®é™…ä¸ŠæŒ‡çš„æ˜¯åˆ†æˆ3ä¸ªæ–‡ä»¶ã€‚
* -AæŒ‡å®šè¾“å‡ºçš„æ–‡ä»¶å

æ ¹æ®Cell Rangerè¯´æ˜ä¹¦å¯¹æ‰¹é‡å¤„ç†å¾—åˆ°çš„ä¸‰ä¸ªæ–‡ä»¶æ›´æ”¹åå­—æ–¹ä¾¿åç»­åˆ†æï¼š
```bash
# æ¯”å¦‚ï¼Œå°†åŸæ¥çš„SRR7692286_1.fastq.gzæ”¹æˆSRR7692286_S1_L001_I1_001.fastq.gz
# ä¾æ¬¡ç±»æ¨ï¼Œå°†åŸæ¥_2çš„æ”¹æˆR1ï¼Œå°†_3æ”¹æˆR2
cat SRR_Acc_List-9245-3.txt | while read i ;do (mv ${i}_1*.gz ${i}_S1_L001_I1_001.fastq.gz;mv ${i}_2*.gz ${i}_S1_L001_R1_001.fastq.gz;mv ${i}_3*.gz ${i}_S1_L001_R2_001.fastq.gz);done
```
![](images/2021-10-22-22-00-56.png)
å…¶ä¸­ï¼Œ<mark>I1ä¸ºindexï¼ŒR1æ˜¯barcodeå’ŒUMIï¼ŒR2æ˜¯æµ‹åºreadï¼Œæ˜¯æœ€é•¿çš„ã€‚</mark>

* **ä¸ºä»€ä¹ˆ10xå•ç»†èƒè½¬å½•ç»„è¡¨è¾¾çŸ©é˜µæœ‰3ä¸ªæ–‡ä»¶**

    å› ä¸º10xå•ç»†èƒè½¬å½•ç»„è¡¨è¾¾çŸ©é˜µé‡Œé¢çš„0å€¼éå¸¸å¤šï¼Œæ‰€ä»¥æ¢æˆ3ä¸ªæ–‡ä»¶å­˜å‚¨æ›´èŠ‚çœç©ºé—´ã€‚

### Cell Ranger
![](images/2021-10-23-09-25-49.png)
å®ƒä¸»è¦åŒ…æ‹¬å››ä¸ªä¸»è¦åŸºå› è¡¨è¾¾åˆ†ææµç¨‹ï¼š
* `mkfastq` ï¼š å®ƒå€Ÿé‰´äº†Illuminaçš„`bcl2fastq` ï¼Œå¯ä»¥å°†ä¸€ä¸ªæˆ–å¤šä¸ªlaneä¸­çš„æ··æ ·æµ‹åºæ ·æœ¬æŒ‰ç…§indexæ ‡ç­¾ç”Ÿæˆæ ·æœ¬å¯¹åº”çš„fastqæ–‡ä»¶
* `count`ï¼šåˆ©ç”¨`mkfastq`ç”Ÿæˆçš„fqæ–‡ä»¶ï¼Œè¿›è¡Œæ¯”å¯¹(åŸºäºSTAR)ã€è¿‡æ»¤ã€UMIè®¡æ•°ã€‚åˆ©ç”¨ç»†èƒçš„barcodeç”Ÿæˆgene-barcodeçŸ©é˜µï¼Œç„¶åè¿›è¡Œæ ·æœ¬åˆ†ç¾¤ã€åŸºå› è¡¨è¾¾åˆ†æã€‚
* `aggr` ï¼šæ¥å—cellranger countçš„è¾“å‡ºæ•°æ®ï¼Œå°†åŒä¸€ç»„çš„ä¸åŒæµ‹åºæ ·æœ¬çš„è¡¨è¾¾çŸ©é˜µæ•´åˆåœ¨ä¸€èµ·ï¼Œæ¯”å¦‚tumorç»„åŸæ¥æœ‰4ä¸ªæ ·æœ¬ï¼ŒPBMCç»„æœ‰ä¸¤ä¸ªæ ·æœ¬ï¼Œç°åœ¨å¯ä»¥ä½¿ç”¨aggrç”Ÿæˆæœ€åçš„tumorå’ŒPBMCä¸¤ä¸ªçŸ©é˜µï¼Œå¹¶ä¸”è¿›è¡Œæ ‡å‡†åŒ–å»æ‰æµ‹åºæ·±åº¦çš„å½±å“
* `reanalyze` ï¼šæ¥å—cellranger `count`æˆ–cellranger `aggr`ç”Ÿæˆçš„gene-barcodeçŸ©é˜µï¼Œä½¿ç”¨ä¸åŒçš„å‚æ•°è¿›è¡Œé™ç»´ã€èšç±»

å®ƒçš„ç»“æœä¸»è¦æ˜¯åŒ…å«æœ‰ç»†èƒä¿¡æ¯çš„BAM, MEX, CSV, HDF5 and HTMLæ–‡ä»¶
#### Cell Rangerè½¯ä»¶å®‰è£…
Cell Rangeré»˜è®¤åœ¨æœ¬åœ°è¿è¡Œ(æˆ–è€…ä½¿ç”¨`--jobmode=local`æŒ‡å®š)ï¼Œå®ƒä¼šå ç”¨90%çš„ç©ºä½™å†…å­˜ä»¥åŠæ‰€æœ‰ç©ºä½™çš„CPUã€‚å¦‚æœè¦è¿›è¡Œèµ„æºé™åˆ¶ï¼Œå¯ä»¥ä½¿ç”¨`â€”localmem`æˆ–è€…`--localcores`
```bash
# 2.0ç‰ˆæœ¬ä¸‹è½½(732M)
curl -o cellranger-2.0.2.tar.gz "http://cf.10xgenomics.com/releases/cell-exp/cellranger-2.0.2.tar.gz?Expires=1557256518&Policy=eyJTdGF0ZW1lbnQiOlt7IlJlc291cmNlIjoiaHR0cDovL2NmLjEweGdlbm9taWNzLmNvbS9yZWxlYXNlcy9jZWxsLWV4cC9jZWxscmFuZ2VyLTIuMC4yLnRhci5neiIsIkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiQVdTOkVwb2NoVGltZSI6MTU1NzI1NjUxOH19fV19&Signature=HoJUuPo4iTFdQgzFU1GH7uKf3uGitQxTjB6WOA9qGPlejf7tNcBPjO65WuSUZ~w8WWdeAvky-oV7XGfheY-bUr2b7QHr7jQEqc84cyU~PLvT~fYjkgC7cG7nlpbJOT~b7U~YH9amvR~SCLlyynp7scPDIA~9~keCYrIPgevTf2QyktybuSyjNTwugefOic~~XFkc9lrS~WQ9MNA1CLl4ExlQKsxWS77PEB6mwrMZXX65obDnZW9fIs3dIny6H5YoadbkgmsT52jmLien6PsG1g2jpAO90pPuHoru8LL64Q9gmB3I0nJAqi3EmrO3GKnUpHUhGb6doKmjSN6XccpmsA__&Key-Pair-Id=APKAI7S6A5RYOXBWRPDA"
# 2.1ç‰ˆæœ¬ä¸‹è½½
curl -o cellranger-2.1.1.tar.gz "http://cf.10xgenomics.com/releases/cell-exp/cellranger-2.1.1.tar.gz?Expires=1557260110&Policy=eyJTdGF0ZW1lbnQiOlt7IlJlc291cmNlIjoiaHR0cDovL2NmLjEweGdlbm9taWNzLmNvbS9yZWxlYXNlcy9jZWxsLWV4cC9jZWxscmFuZ2VyLTIuMS4xLnRhci5neiIsIkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiQVdTOkVwb2NoVGltZSI6MTU1NzI2MDExMH19fV19&Signature=RNQd-gTASTQhtnUSBfQWrnqo6Pyy2wDXtV5tlxkG97727GvoRhMqFXbEsz4gJl2BMckdVvW3S1tZRwRo5pmxPzmhq-8RKxf99pGqlzo84HYqhbIRkxXlIbLbj-u3PUJqo8cesWpbSVSKkS2TCNS-9GMFNieQswqMS2-DN4BqoBOJnWr7T4wlOMd9hypXWwOsW2P2fqaM-WP2ooMyo-oIxm3y9gDghXdDEP5lvHU7GCQcFGGexkdIrD6S5p8JPJ1DB5XieGrtEuP1YVp6tLMGXFoRWXS8dQLI1egWDYlOuRaiQgLIb3o5ZxBg5NpzLPP5kDHMAVzJFdBpf~~rkyNYTA__&Key-Pair-Id=APKAI7S6A5RYOXBWRPDA"
### è§£å‹
tar zxvf cellranger-2.0.2.tar.gz
```
#### å‚è€ƒåŸºå› ç»„åºåˆ—çš„ä¸‹è½½
1. æ ¹æ®éœ€æ±‚ä¸‹è½½åŸºå› ç»„æ³¨é‡Šæ–‡ä»¶
ä»¥ä¸‹æ–‡ä»¶æ˜¯åŸºäºenseblæ•°æ®åº“çš„hg38äººç±»åŸºå› ç»„æ³¨é‡Šæ–‡ä»¶ï¼Œä½†æ˜¯ä¸èƒ½ç›´æ¥ä½¿ç”¨ç½‘ç«™ä¸‹è½½çš„åŸºå› ç»„ä¸æ³¨é‡Šæ–‡ä»¶ï¼Œéœ€è¦è¿‡æ»¤ä¸€ä¸‹ã€‚
```bash
curl -O http://cf.10xgenomics.com/supp/cell-exp/refdata-cellranger-GRCh38-1.2.0.tar.gz
# ç„¶åè§£å‹
tar -xzvf refdata-cellranger-GRCh38-1.2.0.tar.gz
```
2. è‡ªå·±å°è¯•æ„å»º(é‡ç‚¹å’Œéš¾ç‚¹)
ä½†æ˜¯å¾ˆå¤šæ—¶å€™ï¼Œæˆ‘ä»¬éœ€è¦æ ¹æ®è‡ªå·±çš„éœ€è¦ï¼Œè‡ªå®šä¹‰ä¸€å¥—å‚è€ƒä¿¡æ¯ï¼Œä½†éœ€è¦æ³¨æ„ä»¥ä¸‹é—®é¢˜ï¼š
* å‚è€ƒåºåˆ—åªèƒ½æœ‰å¾ˆå°‘çš„ overlapping gene annotationsï¼Œå› ä¸ºreadsæ¯”å¯¹åˆ°å¤šä¸ªåŸºå› ä¼šå¯¼è‡´æµç¨‹æ£€æµ‹çš„åˆ†å­æ•°æ›´å°‘(å®ƒåªè¦uniquely mappedçš„ç»“æœ)
* FASTAä¸GTFæ¯”å¯¹å’ŒSTARå…¼å®¹ï¼ŒGTFæ–‡ä»¶çš„ç¬¬ä¸‰åˆ—ï¼ˆfeature typeï¼‰å¿…é¡»æœ‰exon
```bash
# ä¸‹è½½åŸºå› ç»„
wget ftp://ftp.ensembl.org/pub/release-84/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz
gunzip Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz
# ä¸‹è½½æ³¨é‡Š
wget ftp://ftp.ensembl.org/pub/release-84/gtf/homo_sapiens/Homo_sapiens.GRCh38.84.gtf.gz
gunzip Homo_sapiens.GRCh38.84.gtf.gz
# è½¯ä»¶æ„å»ºæ³¨é‡Š
# é¦–å…ˆåˆ©ç”¨mkgtfè¿‡æ»¤GTFæ–‡ä»¶
# mkgtf <input_gtf> <output_gtf> [--attribute=KEY:VALUE...]
cellranger mkgtf Homo_sapiens.GRCh38.84.gtf Homo_sapiens.GRCh38.84.filtered.gtf \
                --attribute=gene_biotype:protein_coding \
                --attribute=gene_biotype:lincRNA \
                --attribute=gene_biotype:antisense \
                --attribute=gene_biotype:IG_LV_gene \
                --attribute=gene_biotype:IG_V_gene \
                --attribute=gene_biotype:IG_V_pseudogene \
                --attribute=gene_biotype:IG_D_gene \
                --attribute=gene_biotype:IG_J_gene \
                --attribute=gene_biotype:IG_J_pseudogene \
                --attribute=gene_biotype:IG_C_gene \
                --attribute=gene_biotype:IG_C_pseudogene \
                --attribute=gene_biotype:TR_V_gene \
                --attribute=gene_biotype:TR_V_pseudogene \
                --attribute=gene_biotype:TR_D_gene \
                --attribute=gene_biotype:TR_J_gene \
                --attribute=gene_biotype:TR_J_pseudogene \
                --attribute=gene_biotype:TR_C_gene

# gene_biotypeï¼ˆä¹Ÿå°±æ˜¯åŸºå› çš„ç”Ÿç‰©ç±»å‹ï¼‰çš„é”®å€¼å¯¹
$ cat Homo_sapiens.GRCh38.84.filtered.gtf |grep -v "#" |awk -v FS='gene_biotype ' 'NF>1{print $2}'|awk -F ";" '{print $1}'|sort | uniq -c

   213 "IG_C_gene"
    33 "IG_C_pseudogene"
   152 "IG_D_gene"
    76 "IG_J_gene"
     9 "IG_J_pseudogene"
  1209 "IG_V_gene"
   646 "IG_V_pseudogene"
   125 "TR_C_gene"
    16 "TR_D_gene"
   316 "TR_J_gene"
    12 "TR_J_pseudogene"
   848 "TR_V_gene"
   110 "TR_V_pseudogene"
 45662 "antisense"
 58181 "lincRNA"
2337766 "protein_coding"

# åˆ©ç”¨mkrefæ„å»ºå‚è€ƒç´¢å¼•ï¼Œè½¯ä»¶åˆ©ç”¨æ„å»ºå¥½çš„æ³¨é‡Šï¼Œå»æ„å»ºéœ€è¦çš„åŸºå› ç»„
cellranger mkref --genome=GRCh38 \
                --fasta=Homo_sapiens.GRCh38.dna.primary_assembly.fa \
                --genes=Homo_sapiens.GRCh38.84.filtered.gtf \
                --ref-version=1.2.0
```
#### mkfastq æ‹†åˆ†æ•°æ®
<mark>ä¸€èˆ¬æ¥è¯´ï¼Œè¿™ä¸ªæ­¥éª¤æˆ‘ä»¬ä¸ä¼šä½¿ç”¨ï¼Œä»æ•°æ®åº“ä¸‹è½½å¯ä»¥å¾—åˆ°fastqæ–‡ä»¶ï¼Œè€Œä¸”æ•°æ®åº“ä¸€èˆ¬ä¸ä¼šç»™å‡ºBCLsæ ¼å¼çš„åŸå§‹æ–‡ä»¶ã€‚
ç›®çš„ï¼š<mark>å°†æ¯ä¸ªflowcell çš„Illumina sequencer's base call files (BCLs)è½¬ä¸ºfastqæ–‡ä»¶</mark>

ç‰¹è‰²ï¼š å®ƒå€Ÿé‰´äº†Illuminaå‡ºå“çš„`bcl2fastq`ï¼Œå¦å¤–å¢åŠ äº†ï¼š
* å°†10X æ ·æœ¬indexåç§°ä¸å››ç§å¯¡æ ¸è‹·é…¸å¯¹åº”èµ·æ¥ï¼Œæ¯”å¦‚A1å­”æ˜¯æ ·æœ¬SI-GA-A1ï¼Œç„¶åå¯¹åº”çš„å¯¡æ ¸è‹·é…¸æ˜¯GGTTTACT, CTAAACGG, TCGGCGTC, and AACCGTAA ï¼Œé‚£ä¹ˆç¨‹åºå°±ä¼šå»indexæ–‡ä»¶ä¸­å°†å­˜åœ¨è¿™å››ç§å¯¡æ ¸è‹·é…¸çš„fastqç»„åˆåˆ°A1è¿™ä¸ªæ ·æœ¬
* æä¾›è´¨æ§ç»“æœï¼ŒåŒ…æ‹¬barcode è´¨é‡ã€æ€»ä½“æµ‹åºè´¨é‡å¦‚Q30ã€R1å’ŒR2çš„Q30ç¢±åŸºå æ¯”ã€æµ‹åºreadsæ•°ç­‰
* å¯ä»¥ä½¿ç”¨10Xç®€åŒ–ç‰ˆçš„æ ·æœ¬ä¿¡æ¯è¡¨
![](images/2021-10-23-09-47-55.png)
```bash
# ç¬¬ä¸€ç§
$ cellranger mkfastq --id=bcl \
                    --run=/path/to/bcl \
                    --samplesheet=samplesheet-1.2.0.csv
                    --jobmode=local 
                    --localcores=20 
                    --localmem=80
# ç¬¬äºŒç§
$ cellranger mkfastq --id=bcl \
                    --run=/path/to/bcl \
                    --csv=simple-1.2.0.csv
                    --jobmode=local 
                    --localcores=20 
                    --localmem=80
# å…¶ä¸­idæŒ‡å®šè¾“å‡ºç›®å½•çš„åç§°
# runæŒ‡çš„æ˜¯ä¸‹æœºçš„åŸå§‹BCLæ–‡ä»¶ç›®å½•
# -â€“samplesheet ï¼šæ ·å“ä¿¡æ¯åˆ—è¡¨--å…±ä¸‰åˆ—ï¼ˆlane id ,sample name ,index name)ï¼Œæ³¨æ„è¦æ§åˆ¶å¥½æ ¸å¿ƒæ•°å’Œå†…å­˜æ•°
```
#### ğŸ”ºcount ç»†èƒå’ŒåŸºå› çš„å®šé‡
è¿™ä¸ªè¿‡ç¨‹æ˜¯æœ€é‡è¦çš„ï¼Œå®ƒå®Œæˆç»†èƒä¸åŸºå› çš„å®šé‡ï¼Œå®ƒå°†æ¯”å¯¹ã€è´¨æ§ã€å®šé‡éƒ½åŒ…è£…äº†èµ·æ¥ã€‚
```bash
# è¿™æ˜¯ç¤ºä¾‹ï¼Œä¸æ˜¯çœŸå®æ•°æ® #
cellranger count --id=sample345 \
                  --transcriptome=/opt/refdata-cellranger-GRCh38-1.2.0 \
                  --fastqs=/home/scRNA/runs/HAWT7ADXX/outs/fastq_path \
                  --sample=mysample \
                  --expect-cells=1000 \
                  --nosecondary
# idæŒ‡å®šè¾“å‡ºæ–‡ä»¶å­˜æ”¾ç›®å½•å
# transcriptomeæŒ‡å®šä¸CellRangerå…¼å®¹çš„å‚è€ƒåŸºå› ç»„
# fastqsæŒ‡å®šmkfastqæˆ–è€…è‡ªå®šä¹‰çš„æµ‹åºæ–‡ä»¶
# sampleè¦å’Œfastqæ–‡ä»¶çš„å‰ç¼€ä¸­çš„sampleä¿æŒä¸€è‡´ï¼Œä½œä¸ºè½¯ä»¶è¯†åˆ«çš„æ ‡å¿—
# expect-cellsæŒ‡å®šå¤ç°çš„ç»†èƒæ•°é‡ï¼Œè¿™ä¸ªè¦å’Œå®éªŒè®¾è®¡ç»“åˆèµ·æ¥
# nosecondary åªè·å¾—è¡¨è¾¾çŸ©é˜µï¼Œä¸è¿›è¡Œåç»­çš„é™ç»´ã€èšç±»å’Œå¯è§†åŒ–åˆ†æ(å› ä¸ºåæœŸä¼šè‡ªè¡Œç”¨RåŒ…å»åš)
```
è¾“å‡ºæ–‡ä»¶ï¼š
* web_summary.htmlï¼šå®˜æ–¹è¯´æ˜ summary HTML file 
* metrics_summary.csvï¼šCSVæ ¼å¼æ•°æ®æ‘˜è¦
* possorted_genome_bam.bamï¼šæ¯”å¯¹æ–‡ä»¶
* possorted_genome_bam.bam.baiï¼šç´¢å¼•æ–‡ä»¶
* <mark>filtered_gene_bc_matricesï¼šæ˜¯é‡è¦çš„ä¸€ä¸ªç›®å½•ï¼Œä¸‹é¢åˆåŒ…å«äº† barcodes.tsv.gzã€features.tsv.gzã€matrix.mtx.gzï¼Œæ˜¯ä¸‹æ¸¸Seuratã€Scaterã€Monocleç­‰åˆ†æçš„è¾“å…¥æ–‡ä»¶</mark>
* filtered_feature_bc_matrix.h5ï¼šè¿‡æ»¤æ‰çš„barcodeä¿¡æ¯HDF5 format
* raw_feature_bc_matrixï¼šåŸå§‹barcodeä¿¡æ¯
* raw_feature_bc_matrix.h5ï¼šåŸå§‹barcodeä¿¡æ¯HDF5 format
* <mark>analysisï¼šæ•°æ®åˆ†æç›®å½•ï¼Œä¸‹é¢åˆåŒ…å«èšç±»clusteringï¼ˆæœ‰graph-based & k-meansï¼‰ã€å·®å¼‚åˆ†ædiffexpã€ä¸»æˆåˆ†çº¿æ€§é™ç»´åˆ†æpcaã€éçº¿æ€§é™ç»´tsne</mark>
* molecule_info.h5ï¼šä¸‹é¢è¿›è¡Œaggregateä½¿ç”¨çš„æ–‡ä»¶
* cloupe.cloupeï¼šå®˜æ–¹å¯è§†åŒ–å·¥å…·Loupe Cell Browser è¾“å…¥æ–‡ä»¶
#### aggrå¤šä¸ªæ–‡åº“çš„æ•´åˆ
å½“å¤„ç†å¤šä¸ªç”Ÿç‰©å­¦æ ·æœ¬æˆ–è€…ä¸€ä¸ªæ ·æœ¬å­˜åœ¨å¤šä¸ªé‡å¤/æ–‡åº“æ—¶ï¼Œæœ€å¥½çš„æ“ä½œå°±æ˜¯å…ˆåˆ†åˆ«å¯¹æ¯ä¸ªæ–‡åº“è¿›è¡Œå•ç‹¬çš„countå®šé‡ï¼Œç„¶åå°†å®šé‡ç»“æœåˆ©ç”¨aggrç»„åˆèµ·æ¥ã€‚
1.  å¾—åˆ°countç»“æœ
2.  æ„å»ºAggregation CSV
```bash
# AGG123_libraries.csv
library_id,molecule_h5
LV123,/opt/runs/LV123/outs/molecule_info.h5
LB456,/opt/runs/LB456/outs/molecule_info.h5
LP789,/opt/runs/LP789/outs/molecule_info.h5
# å…¶ä¸­
# molecule_h5ï¼šæ–‡ä»¶molecule_info.h5 fileçš„è·¯å¾„
```
3. è¿è¡Œaggr
```bash
cellranger aggr --id=AGG123 \
                 --csv=AGG123_libraries.csv \
                 --normalize=mapped
# ç»“æœè¾“å‡ºåˆ°AGG123è¿™ä¸ªç›®å½•ä¸­
```
### å…³äºä¸Šæ¸¸åˆ†æå¾—åˆ°çš„æ•°æ®é›†
e.g.10X Genomics å…è´¹æä¾›çš„å¤–å‘¨è¡€å•æ ¸ç»†èƒ (PBMC) æ•°æ®é›†
![](images/2021-10-24-21-32-14.png)
* barcodes.tsvï¼šæ¯ä¸ªç»†èƒçš„ç¼–ç ï¼Œç”±ATCGå››ä¸ªç¢±åŸºæ’åˆ—ç»„åˆå½¢æˆçš„ä¸åŒçš„14ä¸ªç¢±åŸºç»„åˆ
* genes.tsvï¼šåŸºå› çš„ensembl ID å’Œgene symbol
![](images/2021-10-24-21-32-37.png)
* matrix.mtxï¼šæ¯ä¸ªç»†èƒä¸åŒåŸºå› æ‰€æµ‹å¾—çš„è¡¨è¾¾çŸ©é˜µï¼Œç¬¬ä¸€åˆ—æ˜¯åŸºå› IDï¼Œä¸genes.tsvè¿›è¡Œå¯¹åº”è½¬æ¢ï¼›ç¬¬äºŒåˆ—æ˜¯ç»†èƒçš„ç¼–å·ï¼Œä¸barcodes.tsvè¿›è¡ŒåŒ¹é…ï¼›ç¬¬ä¸‰åˆ—æ˜¯åŸºå› çš„è¡¨è¾¾é‡ï¼ˆTPMï¼‰ã€‚

## æ•°æ®é¢„å¤„ç†å·¥ä½œæµç¨‹
**åŠ è½½SeuratåŒ…å’Œè¯»å–æ•°æ®**

ä½¿ç”¨Read10X()å‡½æ•°è¯»å–æ•°æ®æ–‡ä»¶ï¼Œå¹¶æ„å»ºæˆ<mark>è¡Œåä¸ºåŸºå› åï¼Œåˆ—åä¸ºç»†èƒçš„barcodeçš„ (UMI) è®¡æ•°çŸ©é˜µ</mark>ï¼Œå…¶ä¸­å¯å‚¨å­˜å•ç»†èƒæ•°æ®é›†çš„ç¨€ç–çŸ©é˜µå’Œåç»­åˆ†æç»“æœã€‚
```r
# BiocManager::install("Seurat")
library(dplyr)
library(Seurat)
library(patchwork)

# Load the PBMC dataset
pbmc.data<-Read10X(data.dir= "../data/pbmc3k/filtered_gene_bc_matrices/hg19/")
# Initialize the Seurat object with the raw (non-normalized data).
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc3k", min.cells = 3, min.features = 200)
pbmc
```
### QCè´¨é‡æ£€æµ‹
> QCè´¨é‡æŒ‡æ ‡ï¼š
* æ¯ä¸ªç»†èƒæ£€æµ‹åˆ°çš„åŸºå› æ•°ç›®
  * ä½è´¨é‡ç»†èƒæˆ–ç©ºæ¶²æ»´é€šå¸¸åªæœ‰å¾ˆå°‘çš„åŸºå› 

  * ç»†èƒåŒè”ä½“æˆ–å¤šé‡è”ä½“å¯èƒ½è¡¨ç°å‡ºå¼‚å¸¸é«˜çš„åŸºå› è®¡æ•°
* æ¯ä¸ªç»†èƒä¸­æ£€æµ‹åˆ°çš„åˆ†å­æ€»æ•°
* ä½è´¨é‡/å‚æ­»çš„ç»†èƒé€šå¸¸è¡¨ç°å‡ºå¹¿æ³›çš„çº¿ç²’ä½“æ±¡æŸ“ï¼Œæ‰€ä»¥ä½¿ç”¨è¯¥PercentageFeatureSet()å‡½æ•°è®¡ç®—çº¿ç²’ä½“ QC æŒ‡æ ‡ï¼Œå…¶ä¸­MT-å¼€å¤´çš„åŸºå› è®¤ä¸ºæ˜¯çº¿ç²’ä½“åŸºå› 
```r
 ###### ç»†èƒè´¨æ§
  ### è®¡ç®—è´¨æ§æŒ‡æ ‡
  # è®¡ç®—ç»†èƒä¸­çº¿ç²’ä½“åŸºå› æ¯”ä¾‹
  pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")
  # è´¨æ§å‰å°æç´å›¾
  plots = list()
  for(i in seq_along(plot.featrures)){
    plots[[i]] = VlnPlot(pbmc, group.by=group, pt.size = 0,
   features = plot.featrures[i]) + theme.set2 + NoLegend()}
  violin <- wrap_plots(plots = plots, nrow=2)    
 ggsave("QC/vlnplot_before_qc.pdf", plot = violin, width = 9, height = 8) 
```
å› æ­¤ï¼Œæˆ‘ä»¬é€šè¿‡å°†QCæŒ‡æ ‡å¯è§†åŒ–ï¼Œå¹¶ä½¿ç”¨è¿™äº›æŒ‡æ ‡æ¥è¿‡æ»¤ä¸ç¬¦åˆæŒ‡æ ‡çš„ç»†èƒã€‚<br>

![](images/2021-10-24-21-49-22.png)

<br>ä»å›¾ä¸­æˆ‘ä»¬å¯ä»¥è§‚å¯Ÿåˆ°nFeature_RNAï¼ŒnCount_RNAå’Œpercent.mtçš„æ•°é‡å¤§éƒ¨åˆ†éƒ½åœ¨å°æç´å›¾çš„åº•éƒ¨ï¼Œå› æ­¤éœ€è¦è¿‡æ»¤ä½è´¨é‡çš„ç»†èƒï¼Œå…¶ä¸­nCount_RNAä¸ºæ¯ä¸ªç»†èƒçš„UMIæ•°ç›®ï¼ŒnFeature_RNAä¸ºæ¯ä¸ªç»†èƒæ‰€æ£€æµ‹åˆ°çš„åŸºå› æ•°ç›®ã€‚
<br><mark>é€šå¸¸ï¼Œæˆ‘ä»¬è¿‡æ»¤å…·æœ‰è¶…è¿‡ 2,500 æˆ–å°‘äº 200 çš„åŸºå› æ•°ç›®çš„ç»†èƒå’Œçº¿ç²’ä½“è®¡æ•° > 5% çš„ç»†èƒã€‚</mark>

```r
### è´¨æ§
  # è®¾ç½®è´¨æ§æ ‡å‡†
  minGene=500
  maxGene=2500
  maxUMI=22000
  pctMT=5
  # æ•°æ®è´¨æ§å¹¶ç»˜åˆ¶å°æç´å›¾
pbmc <- subset(pbmc, subset = nFeature_RNA > minGene & nFeature_RNA < maxGene & nCount_RNA < maxUMI & percent.mt < pctMT )
plots = list()
for(i in seq_along(plot.featrures)){
plots[[i]] = VlnPlot(pbmc, group.by=group, pt.size = 0.01,features = plot.featrures[i]) + theme.set2 +NoLegend()}
violin <- wrap_plots(plots = plots, nrow=1)    
ggsave("QC/vlnplot_after_qc.pdf", plot = violin, width = 9, height = 8)
```
![](images/2021-10-24-21-53-23.png)
å…³äºæ ¸ç³–ä½“å’Œçº¢ç»†èƒçš„è´¨æ§æ ¹æ®éœ€æ±‚é€‰æ‹©æ˜¯å¦éœ€è¦ï¼Œè‹¥éœ€è¦ï¼Œåœ¨ä¸Šä¸€æ­¥ä»£ç æ·»åŠ æ–°çš„è´¨æ§æŒ‡æ ‡ã€‚
```r
# è®¡ç®—ç»†èƒä¸­æ ¸ç³–ä½“åŸºå› æ¯”ä¾‹
  pbmc[["percent.rb"]] <- PercentageFeatureSet(pbmc, pattern = "^RP[LS]")
# è®¡ç®—çº¢ç»†èƒæ¯”ä¾‹ï¼Œæ£€æµ‹æ˜¯å¦æœ‰è¢«è¡€æ±¡æŸ“
  HB.genes <- c("HBA1","HBA2","HBB","HBD","HBE1","HBG1","HBG2","HBM","HBQ1","HBZ")
  HB.genes <- CaseMatch(HB.genes, rownames(pbmc))
  pbmc[["percent.HB"]]<-PercentageFeatureSet(pbmc, features=HB.genes) 
```
> æ£€æµ‹doublets
* Doubletsï¼šå•ä¸ªæ¶²æ»´(droplet)æ•è·ä¸€ä¸ªæ¡å½¢ç ç (barcode bead)å’Œå¤šä¸ªç»†èƒæ ¸ï¼Œå¯¼è‡´äº†å¼‚å¸¸é«˜çš„åŸºå› è®¡æ•°ã€‚
* <mark>ï¼ï¼ï¼</mark>ä¸€èˆ¬æ¥è¯´ï¼Œå»ºè®®æ£€æµ‹doubletsåœ¨QCçš„è´¨æ§nFeature_RNAæ­¥éª¤ä¹‹å‰ï¼Œå…ˆå»é™¤å¼‚å¸¸é«˜çš„åŸºå› è®¡æ•°å¯¹åç»­QCæ•ˆæœæ›´å¥½ã€‚ä¸”ä¸€èˆ¬åœ¨èšç±»æ­¥éª¤æˆ–è€…ç»†èƒç±»å‹æ³¨é‡Šä¹‹åå»é™¤doubletsæ•ˆæœæ›´å¥½ï¼Œå› ä¸ºè¿™æ ·èƒ½ä»å›¾ä¸­æ¸…æ¥šçœ‹åˆ°å“ªä¸€äº›ç»†èƒæ˜¯åœ¨ç°‡çš„è¾¹ç¼˜å’Œå±äºå“ªä¸€ç±»çš„ç»†èƒï¼Œè€Œè¿™ä¸€äº›è¾¹ç¼˜çš„ç»†èƒä¸€èˆ¬æ¥è¯´æ˜¯doubletsã€‚
* <mark>å»é™¤doubletsä¹‹åï¼Œå†å¯¹å…¶ä½™çš„åŸå§‹countsçŸ©é˜µè¿›è¡Œæ–°ä¸€è½®çš„Seuratçš„æ ‡å‡†åˆ†ææµç¨‹ã€‚</mark>
1. DoubletFinderåŒ…
*ï¼ˆä¸Seuratçš„äº¤äº’ï¼šåœ¨seuratæ ‡å‡†æµç¨‹è¿›è¡Œåˆ°t-SNEå’ŒUMAPé™ç»´ä¹‹åï¼ŒFindAllMarkersä¹‹å‰è¿›è¡ŒDoubletFinderæ“ä½œã€‚ï¼‰*

åŸç†ï¼šä»ç°æœ‰çš„çŸ©é˜µçš„ç»†èƒä¸­æ ¹æ®æˆ‘ä»¬é¢„å…ˆå®šä¹‰å¥½çš„ç»†èƒç±»å‹æ¨¡æ‹Ÿä¸€äº›åŒç»†èƒå‡ºæ¥ï¼ˆæ¯”å¦‚å•æ ¸å’ŒTç»†èƒçš„åŒç»†èƒã€Bç»†èƒå’Œä¸­æ€§ç²’ç»†èƒçš„åŒç»†èƒç­‰ç­‰ï¼‰ï¼Œå°†æ¨¡æ‹Ÿå‡ºçš„åŒç»†èƒå’ŒåŸæœ‰çŸ©é˜µçš„ç»†èƒæ··åˆåœ¨ä¸€èµ·ï¼Œè¿›è¡Œé™ç»´èšç±»ï¼ŒåŸåˆ™ä¸Šäººå·¥æ¨¡æ‹Ÿçš„doubletsä¼šä¸çœŸå®çš„doubletsè·ç¦»è¾ƒè¿‘ï¼Œå› æ­¤è®¡ç®—æ¯ä¸ªç»†èƒKæœ€è¿‘é‚»ç»†èƒä¸­äººå·¥æ¨¡æ‹Ÿdoubletsçš„æ¯”ä¾‹ (pANN)ï¼Œå°±å¯ä»¥æ ¹æ®pANNå€¼å¯¹æ¯ä¸ªbarcodeçš„doubletsæ¦‚ç‡è¿›è¡Œæ’åºã€‚å¦å¤–ä¾æ®æ³Šæ¾åˆ†å¸ƒçš„ç»Ÿè®¡åŸç†å¯ä»¥è®¡ç®—æ¯ä¸ªæ ·æœ¬ä¸­doubletsçš„æ•°é‡ï¼Œç»“åˆä¹‹å‰çš„ç»†èƒpANNå€¼æ’åºï¼Œå°±å¯ä»¥è¿‡æ»¤doubletsäº†ã€‚
```r
devtools::install_github('chris-mcginnis-ucsf/DoubletFinder')
library(DoubletFinder)
  ## æ£€æµ‹doublets
  pc.num=1:10
  # å¯»æ‰¾æœ€ä¼˜pKå€¼
  sweep.res.list <- paramSweep_v3(pbmc, PCs = pc.num, sct = T) #ä½¿ç”¨logæ ‡å‡†åŒ–ï¼Œsctå‚æ•°è®¾ç½®ä¸º sct = Fï¼ˆé»˜è®¤ ï¼‰,å¦‚ä½¿ç”¨SCTæ ‡å‡†åŒ–æ–¹æ³•ï¼Œè®¾ç½®ä¸ºT
  saveRDS(sweep.res.list, "QC/sweep.res.list.rds")
# å½“GTè®¾ç½®ä¸º TRUEæ—¶ï¼Œå¯ç”¨äºROCåˆ†æã€‚ é»˜è®¤è®¾ç½®ä¸º FALSEã€‚
  sweep.stats <- summarizeSweep(sweep.res.list, GT = FALSE)  
  bcmvn <- find.pK(sweep.stats) #å¯ä»¥çœ‹åˆ°æœ€ä½³å‚æ•°çš„ç‚¹
  # æå–æœ€ä½³pkå€¼
  pK_bcmvn <- bcmvn$pK[which.max(bcmvn$BCmetric)] %>% as.character() %>% as.numeric()
  # æ’é™¤ä¸èƒ½æ£€å‡ºçš„åŒæºdoubletsï¼Œä¼˜åŒ–æœŸæœ›çš„doubletsæ•°é‡ï¼ŒåŒç»†èƒæœ‰ä¸¤ç§ï¼ŒåŒæºåŒç»†èƒå’Œå¼‚æºåŒç»†èƒã€‚DoubletFinderåªèƒ½æ£€æµ‹å¼‚æºåŒç»†èƒã€‚æ‰€ä»¥éœ€è¦æŠŠåŒæºåŒç»†èƒå¯èƒ½çš„æ¯”ç‡å»é™¤æ‰ï¼Œä»¥ä¼˜åŒ–æœŸæœ›çš„doubletsæ•°é‡ã€‚
# æŒ‰æ¯å¢åŠ 1000ä¸ªç»†èƒï¼ŒåŒç»†èƒæ¯”ç‡å¢åŠ åƒåˆ†ä¹‹8æ¥è®¡ç®— 
DoubletRate = ncol(pbmc)*8*1e-6 
# ä¼°è®¡åŒæºåŒç»†èƒæ¯”ä¾‹ï¼Œæ ¹æ®modelHomotypic()ä¸­çš„å‚æ•°ä¸ºäººä¸ºæ··åˆåŒç»†èƒã€‚è¿™é‡Œæ˜¯ä»seurat_clustersä¸­æ¥æ··åŒç»†èƒ 
  homotypic.prop <- modelHomotypic(pbmc$SingleR)
# è¯¥éƒ¨åˆ†ä½¿ç”¨SingleRåçš„æ•°æ®æ˜¯ä¸ºäº†å¯è§†åŒ–å“ªäº›ç±»å‹çš„ç»†èƒä¸­æœ‰doublets 
  nExp_poi <- round(DoubletRate*ncol(pbmc))  # è®¡ç®—åŒç»†èƒæ¯”ä¾‹
  nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))  # ä½¿ç”¨åŒæºåŒç»†èƒæ¯”ä¾‹å¯¹è®¡ç®—çš„åŒç»†èƒæ¯”ä¾‹è¿›è¡Œæ ¡æ­£
  # ä½¿ç”¨ç¡®å®šå¥½çš„å‚æ•°é‰´å®šdoublets
  pbmc <- doubletFinder_v3(pbmc, PCs = pc.num, pN = 0.25, pK = pK_bcmvn, nExp = nExp_poi.adj, reuse.pANN = F, sct = T)
# å¯è§†åŒ–ç»“æœ
  names(pbmc@meta.data) = gsub("DF.classifications.*", "DF.classifications", colnames(pbmc@meta.data))  
  p1 <- DimPlot(pbmc, group.by = "DF.classifications") + scale_color_manual(values = c("red", "gray"))
  p2 <- DimPlot(pbmc, label = T)
  p <- p1|p2
  ggsave("QC/Doublets_DFpred.pdf", p, width = 12, height = 5)
```
![](images/2021-10-24-21-59-08.png)
<br>
2. ScrubletåŒ…(python)
åŸç†ï¼šç»™å®šä¸€ä¸ªåŸå§‹çš„ï¼ˆæœªå½’ä¸€åŒ–çš„ï¼‰UMIçŸ©é˜µï¼Œä»¥ç»†èƒä¸ºè¡Œï¼ŒåŸºå› ä¸ºåˆ—çš„çŸ©é˜µcounts_matrixè®¡æ•°ï¼Œè®¡ç®—æ¯ä¸ªå•å…ƒçš„å¤šç»†èƒå¾—åˆ†ã€‚

> æ³¨æ„äº‹é¡¹ï¼š
1. å¤„ç†æ¥è‡ªå¤šä¸ªæ ·æœ¬çš„æ•°æ®æ—¶ï¼Œè¯·åˆ†åˆ«å¯¹æ¯ä¸ªæ ·æœ¬è¿è¡ŒScrubletã€‚ å› ä¸ºScrubletæ—¨åœ¨æ£€æµ‹ç”±ä¸¤ä¸ªç»†èƒçš„éšæœºå…±å°è£…å½¢æˆçš„technical doubletsï¼Œæ‰€ä»¥åœ¨mergedæ•°æ®é›†ä¸Šå¯èƒ½ä¼šè¡¨ç°ä¸ä½³ï¼Œå› ä¸ºç»†èƒç±»å‹æ¯”ä¾‹ä¸ä»£è¡¨ä»»ä½•å•ä¸ªæ ·å“ï¼› 
2. æ£€æŸ¥doublet scoreé˜ˆå€¼æ˜¯å¦åˆç†ï¼Œå¹¶åœ¨å¿…è¦æ—¶è¿›è¡Œæ‰‹åŠ¨è°ƒæ•´ã€‚å¹¶ä¸æ˜¯æ‰€æœ‰æƒ…å†µå‘ä¸‹doublet scoreçš„ç›´æ–¹åˆ†å¸ƒå›¾éƒ½æ˜¯å‘ˆç°æ ‡å‡†çš„åŒå³°ï¼› 
3. UMAPæˆ–t-SNEå¯è§†åŒ–çš„ç»“æœä¸­ï¼Œé¢„æµ‹çš„åŒç»†èƒåº”è¯¥å¤§ä½“ä¸Šå…±å®šä½ï¼ˆå¯èƒ½åœ¨å¤šä¸ªç»†èƒç¾¤ä¸­ï¼‰ã€‚å¦‚æœä¸æ˜¯ï¼Œåˆ™å¯èƒ½éœ€è¦è°ƒæ•´doublet scoreé˜ˆå€¼ï¼Œæˆ–æ›´æ”¹é¢„å¤„ç†å‚æ•°ä»¥æ›´å¥½åœ°è§£ææ•°æ®ä¸­å­˜åœ¨çš„ç»†èƒçŠ¶æ€ã€‚
```python
import scrublet as scr
import scipy.io
import matplotlib.pyplot as plt
import numpy as np
import os
import pandas as pd
input_dir = '.../pbmc3k_filtered_gene_bc_matrices/filtered_gene_bc_matrices/hg19'
counts_matrix = scipy.io.mmread(input_dir + '/matrix.mtx').T.tocsc()
genes = np.array(scr.load_genes(input_dir + '/genes.tsv', delimiter='\t', column=1))
out_df = pd.read_csv(input_dir + '/barcodes.tsv', header = None, index_col=None, names=['barcode'])
# åˆå§‹åŒ–Scrubletå¯¹è±¡ï¼Œexpected_doublet_rateé€šå¸¸ä¸º0.05-0.1ï¼Œç»“æœå¯¹è¯¥å‚æ•°ä¸æ•æ„Ÿ
scrub = scr.Scrublet(counts_matrix, expected_doublet_rate=0.06)
# è®¡ç®—doublet score
doublet_scores, predicted_doublets = scrub.scrub_doublets(min_counts=2, min_cells=3, min_gene_variability_pctl=85, n_prin_comps=30)
```
![](images/2021-10-24-22-00-53.png)
```python
# ç»˜åˆ¶doublet scoreåˆ†å¸ƒç›´æ–¹å›¾
scrub.call_doublets(threshold=0.30)
# å¦‚æœè‡ªåŠ¨é˜ˆå€¼æ£€æµ‹æ•ˆæœä¸ä½³ï¼Œåˆ™å¯ä»¥ä½¿ç”¨call_doublets()å‡½æ•°è°ƒæ•´é˜ˆå€¼
# Doublet scoreåˆ†å¸ƒç›´æ–¹å›¾åŒ…æ‹¬è§‚å¯Ÿåˆ°çš„è½¬å½•ç»„å’Œæ¨¡æ‹Ÿçš„doubletï¼Œæ¨¡æ‹Ÿçš„doubletç›´æ–¹å›¾é€šå¸¸æ˜¯åŒå³°çš„ã€‚
# ç”»doublet scoreç›´æ–¹å›¾
scrub.plot_histogram()
### ç†æƒ³æƒ…å†µä¸‹ï¼Œé˜ˆå€¼åº”åœ¨æ¨¡æ‹Ÿdoubletç›´æ–¹å›¾çš„ä¸¤ç§æ¨¡å¼ä¹‹é—´è®¾ç½®æœ€å°å€¼
```
![](images/2021-10-24-22-01-34.png)
```python
# é™ç»´å¯è§†åŒ–
print('Running UMAP...')
scrub.set_embedding('UMAP', scr.get_umap(scrub.manifold_obs_, 10, min_dist=0.3))
print('Done.')
### UMAPå¯è§†åŒ–
scrub.plot_embedding('UMAP', order_points=True)
```
![](images/2021-10-24-22-02-01.png)
```python
# doubletså æ¯”
print (scrub.detected_doublet_rate_)
# 0.014074074074074074
# æŠŠdoubletsé¢„æµ‹ç»“æœä¿å­˜åˆ°æ–‡ä»¶ï¼Œåç»­ç”¨Seuratç­‰è½¯ä»¶å¤„ç†çš„æ—¶å€™å¯ä»¥å¯¼å…¥doubletsçš„é¢„æµ‹ç»“æœå¯¹barcodeè¿›è¡Œç­›é€‰ã€‚
out_df['doublet_scores'] = doublet_scores
out_df['predicted_doublets'] = predicted_doublets
out_df.to_csv(input_dir + '/doublet.txt', index=False,header=True)
out_df.head()
```
![](images/2021-10-24-22-03-01.png)
### å½’ä¸€åŒ–æ•°æ®
æˆ‘ä»¬é‡‡ç”¨<mark>å…¨å±€ç¼©æ”¾å½’ä¸€åŒ–æ–¹æ³•â€œLogNormalizeâ€</mark>ï¼Œè¯¥æ–¹æ³•å°†æ¯ä¸ªç»†èƒçš„ç‰¹å¾è¡¨è¾¾å¼æµ‹é‡å€¼ä¸æ€»è¡¨è¾¾å¼è¿›è¡Œå½’ä¸€åŒ–ï¼Œä¹˜ä»¥ä¸€ä¸ªæ¯”ä¾‹å› å­ï¼ˆé»˜è®¤ä¸º 10,000ï¼‰ï¼Œå¹¶å¯¹ç»“æœè¿›è¡Œå¯¹æ•°è½¬æ¢ã€‚å°†æœ‰é‡çº²çš„è¡¨è¾¾å¼ï¼Œç»è¿‡å˜æ¢ï¼ŒåŒ–ä¸ºæ— é‡çº²çš„è¡¨è¾¾å¼ï¼Œæ ‡å‡†åŒ–å€¼å­˜å‚¨åœ¨`pbmc[["RNA"]]@data`ã€‚
```r
pbmc <- NormalizeData(pbmc, normalization.method = "LogNormalize", scale.factor = 10000)
```
### è¯†åˆ«é«˜å¯å˜åŸºå› 
* é«˜å¯å˜åŸºå› ï¼šåœ¨æŸäº›ç»†èƒä¸­é«˜åº¦è¡¨è¾¾ï¼Œè€Œåœ¨å…¶ä»–ç»†èƒä¸­è¡¨è¾¾ä½ï¼ˆåœ¨ä¸‹æ¸¸åˆ†æä¸­å…³æ³¨è¿™äº›åŸºå› æœ‰åŠ©äºçªå‡ºå•ç»†èƒæ•°æ®é›†ä¸­çš„ç”Ÿç‰©ä¿¡å·ã€‚ï¼‰
* ä½¿ç”¨å‡å€¼ä¸æ–¹å·®ä¹‹é—´çš„å…³ç³»ï¼Œæ¥æŒ‘é€‰é«˜å˜åŸºå› ï¼Œé»˜è®¤è¿”å›å‰2000ä¸ªé«˜å˜åŸºå› è¿›å…¥ä¸‹æ¸¸åˆ†æï¼Œå¦‚PCAã€‚
```r
pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(pbmc), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(pbmc)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1 + plot2
```
![](images/2021-10-24-22-06-46.png)
### æ ‡å‡†åŒ–æ•°æ®
åœ¨é™ç»´æŠ€æœ¯ï¼ˆå¦‚ PCAï¼‰ä¹‹å‰çš„æ ‡å‡†é¢„å¤„ç†æ­¥éª¤ï¼Œä½¿ç”¨ScaleData()å‡½æ•°ï¼Œä½¿æ•°æ®ç¬¦åˆæ ‡å‡†æ­£æ€åˆ†å¸ƒï¼Œå› ä¸ºPCAåˆ†æé»˜è®¤æ•°æ®æ˜¯æœä»æ­£æ€åˆ†å¸ƒï¼Œå…¶ç»“æœå­˜å‚¨åœ¨pbmc[["RNA"]]@scale.dataã€‚
ScaleData()å‡½æ•°ï¼š

* æ”¹å˜æ¯ä¸ªåŸºå› çš„è¡¨è¾¾ï¼Œä½¿è·¨ç»†èƒçš„å¹³å‡è¡¨è¾¾ä¸º 0
* ç¼©æ”¾æ¯ä¸ªåŸºå› çš„è¡¨è¾¾ï¼Œä½¿ç»†èƒé—´çš„æ–¹å·®ä¸º 1
``` r
all.genes <- rownames(pbmc)
pbmc <- ScaleData(pbmc, features = all.genes)
```
> SCTransform()å‡½æ•°

<mark>å¯ä»¥ä»£æ›¿ä»¥ä¸Šä¸‰ä¸ªå¤„ç†æ­¥éª¤çš„å‡½æ•°ï¼ˆNormalizeData,
ScaleData,FindVariableFeaturesï¼‰çš„è¿è¡Œã€‚</mark>è¿™æ˜¯ä¸€ä¸ªç”¨æ–¹å·®ç¨³å®šå˜æ¢å¯¹å•ç»†èƒUMI count æ•°æ®æ ‡å‡†åŒ–çš„æ–¹æ³•ï¼Œæ–¹å·®ç¨³å®šå˜æ¢æ˜¯åŸºäºè´ŸäºŒé¡¹å›å½’ã€‚è¿™ä¸ªå‡½æ•°åœ¨å¯¹æ•°æ®è¿›è¡Œå‡ä¸€åŒ–çš„åŒæ—¶è¿˜å¯ä»¥å»é™¤çº¿ç²’ä½“çº¢ç»†èƒç­‰æ··æ‚å› ç´ çš„å½±å“ã€‚
ä¼˜ç‚¹ï¼š
* å¯¹æµ‹åºæ·±åº¦çš„æ ¡æ­£æ•ˆæœè¦å¥½äºlogå½’ä¸€åŒ–ã€‚ï¼ˆ10ä¸‡ä»¥å†…çš„ç»†èƒéƒ½å»ºè®®ä½¿ç”¨SCTæ ‡å‡†åŒ–ï¼‰
* SCTransformå¯¹æµ‹åºæ·±åº¦çš„æ ¡æ­£æ•ˆæœå¾ˆå¥½ï¼Œä¹Ÿå¯ç”¨äºæ ¡æ­£çº¿ç²’ä½“ç­‰å› ç´ çš„å½±å“ï¼Œä½†ä¸èƒ½ç”¨äºæ‰¹æ¬¡æ ¡æ­£ã€‚
```r
pbmc <- NormalizeData(pbmc) %>% FindVariableFeatures(nfeatures = 2000) %>% ScaleData(vars.to.regress = "percent.mt"))
 
pbmc <- SCTransform(pbmc, vars.to.regress = "percent.mt")
```
ä»¥ä¸Šä¸¤è¡Œä»£ç åŠŸèƒ½ç›¸åŒã€‚

### æ‰§è¡Œçº¿æ€§é™ç»´

1. PCAä¸»æˆåˆ†åˆ†æ
æ¥ä¸‹æ¥ï¼Œæˆ‘ä»¬å¯¹æ ‡å‡†åŒ–åçš„æ•°æ®è¿›è¡ŒPCAåˆ†æã€‚é»˜è®¤çš„ï¼Œåªæ˜¯ç”¨å‰é¢å†³å®šçš„é«˜å˜åŸºå› è¿›è¡ŒPCAåˆ†æï¼Œä¹Ÿå¯ä»¥ä½¿ç”¨featureså‚æ•°è®¾ç½®ç”¨æˆ·è‡ªå·±é€‰æ‹©çš„åŸºå› è¿›è¡ŒPCAåˆ†æã€‚

```R
pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))
# Examine and visualize PCA results a few different ways
print(pbmc[["pca"]], dims = 1:5, nfeatures = 5)
VizDimLoadings(pbmc, dims = 1:2, reduction = "pca")
DimPlot(pbmc, reduction = "pca")
```
![](images/2021-10-24-22-13-00.png)
![](images/2021-10-24-22-13-18.png)
```r
### DimHeatmap()å…è®¸æˆ‘ä»¬æ¢ç´¢æ•°æ®ä¸­çš„æœ€åˆçš„å¼‚è´¨æ€§ï¼Œç„¶åå†³å®šä¸‹æ¸¸ä½¿ç”¨å¤šå°‘ä¸ªPCè¿›è¡Œåˆ†æã€‚
DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)
DimHeatmap(pbmc, dims = 1:15, cells = 500, balanced = TRUE)
```
![](images/2021-10-24-22-13-39.png)
2. NMFéè´ŸçŸ©é˜µåˆ†è§£
<mark>ï¼ˆä¸€èˆ¬åœ¨PCAåå†ä½¿ç”¨NMFï¼Œä½¿å¾—å› å­è§£é‡Šæ€§æ›´é«˜ï¼‰</mark>
<br>* çŸ©é˜µåˆ†è§£çš„æ–¹æ³•æœ‰å¾ˆå¤šç§ï¼Œå¦‚å¥‡å¼‚å€¼åˆ†è§£ (singular value decomposition, SVD) ã€ç‹¬ç«‹æˆåˆ†åˆ†æ (independent component analysis, ICA) ã€ä¸»æˆåˆ†åˆ†æ (principal component analysis, PCA) ç­‰ã€‚è¿™äº›æ–¹æ³•çš„å…±åŒç‰¹ç‚¹æ˜¯ï¼Œå³ä½¿åˆå§‹çŸ©é˜µVå…ƒç´ æ˜¯éè´Ÿçš„ï¼Œåˆ†è§£å‡ºæ¥çš„å› å­ W å’Œ H ä¸­çš„å…ƒç´ å¾€å¾€å«æœ‰è´Ÿå€¼å…ƒç´ ã€‚ä»è®¡ç®—ç§‘å­¦çš„è§’åº¦æ¥çœ‹ï¼Œåˆ†è§£å‡ºæ¥çš„å› å­ W å’Œ H ä¸­çš„å…ƒç´ å«æœ‰è´Ÿå€¼å…ƒç´ å¹¶æ²¡æœ‰é—®é¢˜ï¼Œ ä½†è´Ÿå€¼å…ƒç´ é€šå¸¸æ˜¯æ— æ³•è§£é‡Šçš„ã€‚
<br>* éè´ŸçŸ©é˜µåˆ†è§£(Non-negative Matrix Factorization, NMF)æœ¬è´¨ä¸Šè¯´æ˜¯ä¸€ç§çŸ©é˜µåˆ†è§£çš„æ–¹æ³•ï¼Œæœ€é‡è¦çš„ç‰¹ç‚¹æ˜¯éè´Ÿæ€§çº¦æŸ,å¯¹äºä»»æ„ç»™å®šçš„ä¸€ä¸ªéè´ŸçŸ©é˜µVï¼ŒNMFç®—æ³•èƒ½å¤Ÿå¯»æ‰¾åˆ°ä¸€ä¸ªéè´ŸçŸ©é˜µWå’Œä¸€ä¸ªéè´ŸçŸ©é˜µHï¼Œä½¿å¾— Vâ‰ˆW*Hæˆç«‹ ï¼Œä»è€Œå°†ä¸€ä¸ªéè´Ÿçš„çŸ©é˜µåˆ†è§£ä¸ºå·¦å³ä¸¤ä¸ªéè´ŸçŸ©é˜µçš„ä¹˜ç§¯ã€‚
<br>* ç›¸æ¯”äºPCAçš„ä¼˜åŠ¿ï¼šè™½ç„¶NMFè¿è¡Œçš„æ—¶é—´æ¯”PCAæ—¶é—´æ›´é•¿ï¼Œä½†å®ƒåˆ†è§£çš„å› å­å¾ˆå®¹æ˜“ä¸ç»†èƒç±»å‹æˆ–è¡¨è¾¾æ¨¡å¼å¯¹åº”èµ·æ¥ï¼Œå³NMFçš„å› å­å¯è§£é‡Šæ€§æ›´å¼ºã€‚

```r
### å…³äºNMFçš„å‡½æ•°
nmf(x, rank, method, seed, nrun, ...)
xï¼šå¾…åˆ†è§£éè´ŸçŸ©é˜µï¼Œæ•°æ®æ ¼å¼å¯ä»¥æ˜¯matrixï¼Œdata.frameï¼Œ ExpressionSet
rankï¼šåˆ†è§£çš„åŸºæ•°é‡ï¼Œå¯¹äºå•ç»†èƒæ•°æ®ï¼Œå¯ä»¥è®¾ç½®ä¸ºæœŸæœ›çš„ç»†èƒç±»å‹æ•°é‡æˆ–è¡¨è¾¾æ¨¡å¼æ•°é‡
methodï¼šå› å¼åˆ†è§£çš„å¸¸ç”¨æ–¹æ³•ï¼Œè¿™é‡Œä»‹ç»ä¸‰ç§å¸¸ç”¨çš„
        1ã€åŸºäºKL æ•£åº¦è¿›è¡Œåº¦é‡ç›®æ ‡å‡½æ•°çš„å¤šé‡è¿­ä»£æ¢¯åº¦ä¸‹é™ç®—æ³•â€”â€”brunet(é»˜è®¤ç®—æ³•)
        2ã€åŸºäºæ¬§å‡ é‡Œå¾—è·ç¦»åº¦é‡ç›®æ ‡å‡½æ•°çš„å¤šé‡è¿­ä»£æ¢¯åº¦ä¸‹é™ç®—æ³•â€”â€”lee
        3ã€äº¤æ›¿æœ€å°äºŒä¹˜æ³•(Alternating Least Squares(ALS))â€”â€”snmf/r  
seedï¼šå› å¼åˆ†è§£çš„åˆå§‹åŒ–ç§å­    
nrunï¼šè¿è¡Œæ¬¡æ•° 
è¿è¡Œä»£ç ï¼š
# å®‰è£…NMFåŸºç¡€åŒ…
BiocManager::install('Biobase')
install.packages('NMF')
library(Seurat)
library(tidyverse)
library(NMF)
rm(list = ls())

pbmc <- NormalizeData(pbmc) %>% FindVariableFeatures() %>% ScaleData(do.center = F)###é‡æ–°è¿›è¡Œæ ‡å‡†åŒ–å’Œå½’ä¸€åŒ–ï¼Œè®¾ç½®do.center = Få¯ä»¥ä¸ä¼šå‡ºç°è´Ÿå€¼
vm <- pbmc@assays$RNA@scale.data
saveRDS(vm, file = "pbmc_vm.rds")
res <- nmf(vm, 12, method = "snmf/r")#å¾ˆæ…¢ï¼Œrankå€¼é€‰æ‹©æ¯”ç›®çš„é¢„æœŸçš„ç»†èƒç±»å‹/ç»†èƒçŠ¶æ€ç¨å¤§çš„å€¼ï¼Œå› ä¸ºåˆ†è§£çš„ä¸€äº›å› å­å›å¤å³åˆ°çº¿ç²’ä½“æ ¸ç³–ä½“ç­‰å™ªéŸ³ï¼Œè€Œä¸ä¼šè½åˆ°ä¸€ä¸ªå…·ä½“çš„ç»†èƒäºšç¾¤ä¸Š
save(res, file = "pbmc_nmf_res.rda" )
## åˆ†è§£ç»“æœè¿”å›sueratå¯¹è±¡
pbmc@reductions$nmf <- pbmc@reductions$pca
pbmc@reductions$nmf@cell.embeddings <- t(coef(res))    
pbmc@reductions$nmf@feature.loadings <- basis(res)  

# æå–åˆ†è§£å¾—åˆ°çš„æ¯ä¸ªå› å­
#æ¯ä¸ªå› å­æå–30ä¸ª
fs <- extractFeatures(res, 30L)
fs <- lapply(fs, function(x) rownames(res)[x])
fs <- do.call("rbind", fs)
rownames(fs) <- paste0("cluster", 1:12)
write.csv(t(fs), "NMF_TopGenes. csv")
DT::datatable(t(fs))

# é€‰æ‹©ç”¨äºåç»­åˆ†æçš„å› å­ï¼Œä½¿ç”¨NMFè¿è¡Œçš„ç»“æœè¿›è¡Œé™ç»´èšç±»
###é€‰æ‹©ç”¨äºåç»­åˆ†æçš„å› å­
s.f <- 1:12
#å› å­1ä¸»è¦æ˜¯çº¿ç²’ä½“å’Œæ ¸ç³–ä½“
##é™ç»´
cell1 <- colnames(pbmc)
cel12 <- colnames(coef(res))
cells <- intersect(cell1, cel12)
pbmc<<-pbmc[,cells]
pbmc <- RunPCA(pbmc,verbose = F)
pbmc@reductions$nmf <- pbmc@reductions$pca
pbmc@reductions$nmf@cell.embeddings <- t(coef(res)[,cells])
pbmc@reductions $nmf@feature.loadings <- basis(res)
pbmc <- RunUMAP(pbmc, reduction= 'nmf',dims=s.f)
##åŸºäºNMFé™ç»´çŸ©é˜µçš„èšç±»
pbmc <- FindNeighbors(pbmc, reduction= 'nmf', dims=s.f) %>% FindClusters()
##åŸºäºå› å­æœ€å¤§è½½è·åˆ†ç±»
pbmc$cluster <- apply(NMF::coefficients(res)[s.f,], 2,which.max)

# é™ç»´èšç±»ç»“æœå¯è§†åŒ–
p1 <- DimPlot(pbmc, label = T) + ggtitle("Clustered by Louvain")
p2 <- DimPlot(pbmc, group.by = "cluster", label = T) + ggtitle("Clustered by maxloading")
pc <- p1|p2
ggsave(" pbmc_NMF_Cluster.pdf", pc, width = 10, height = 5)
```

### ç¡®å®šæ•°æ®é›†çš„â€œç»´åº¦â€
ä¸ºäº†å…‹æœ scRNA-seq æ•°æ®çš„ä»»ä½•å•ä¸ªç‰¹å¾ä¸­çš„å¹¿æ³›æŠ€æœ¯å™ªéŸ³ï¼ŒSeurat æ ¹æ®å…¶ PCA åˆ†æ•°å¯¹ç»†èƒè¿›è¡Œèšç±»ï¼Œæ¯ä¸ª PC åŸºæœ¬ä¸Šä»£è¡¨ä¸€ä¸ªâ€œç»¼åˆç‰¹å¾â€ï¼Œè¯¥â€œç»¼åˆç‰¹å¾â€ç»„åˆäº†ç›¸å…³ç‰¹å¾é›†çš„ä¿¡æ¯ã€‚å› æ­¤ï¼Œé¡¶éƒ¨ä¸»æˆåˆ†ä»£è¡¨æ•°æ®é›†çš„ç¨³å¥å‹ç¼©çš„ä¿¡æ¯ã€‚ä½†æ˜¯ï¼Œæˆ‘ä»¬åº”è¯¥é€‰æ‹©å¤šå°‘ä¸ªPCè¿›è¡Œåˆ†æï¼Ÿ
1. JackStraw ç¨‹åºå¯å‘çš„é‡é‡‡æ ·æµ‹è¯•ã€‚æˆ‘ä»¬éšæœºç½®æ¢æ•°æ®çš„ä¸€ä¸ªå­é›†ï¼ˆé»˜è®¤ä¸º 1%ï¼‰å¹¶é‡æ–°è¿è¡Œ PCAï¼Œæ„å»ºç‰¹å¾åˆ†æ•°çš„â€œé›¶åˆ†å¸ƒâ€ï¼Œå¹¶é‡å¤æ­¤è¿‡ç¨‹ã€‚æˆ‘ä»¬å°†â€œæ˜¾ç€â€PC è¯†åˆ«ä¸ºé‚£äº›å…·æœ‰å¤§é‡ä½ p å€¼ç‰¹å¾çš„ PCã€‚
```r
# NOTE: This process can take a long time for big datasets, comment out for expediency. More
# approximate techniques such as those implemented in ElbowPlot() can be used to reduce
# computation time
pbmc <- JackStraw(pbmc, num.replicate = 100)
pbmc <- ScoreJackStraw(pbmc, dims = 1:20)
```
2. è¯¥JackStrawPlot()å‡½æ•°æä¾›äº†ä¸€ä¸ªå¯è§†åŒ–å·¥å…·ï¼Œç”¨äºå°†æ¯ä¸ª PC çš„ p å€¼åˆ†å¸ƒä¸å‡åŒ€åˆ†å¸ƒï¼ˆè™šçº¿ï¼‰è¿›è¡Œæ¯”è¾ƒã€‚â€œæ˜¾è‘—â€PC å°†æ˜¾ç¤ºå‡ºå¤§é‡å…·æœ‰ä½ p å€¼çš„ç‰¹å¾ï¼ˆè™šçº¿ä¸Šæ–¹çš„å®çº¿ï¼‰ã€‚åœ¨è¿™ç§æƒ…å†µä¸‹ï¼Œåœ¨å‰ 10-12 ä¸ª PC ä¹‹åï¼Œé‡è¦æ€§ä¼¼ä¹æ€¥å‰§ä¸‹é™ã€‚
```r
JackStrawPlot(pbmc, dims = 1:15)
```
![](images/2021-10-24-22-18-04.png)
3. è‚˜å›¾ï¼ˆç¢çŸ³å›¾ï¼‰ï¼šåŸºäºæ¯ä¸ªï¼ˆElbowPlot()å‡½æ•°ï¼‰è§£é‡Šçš„æ–¹å·®ç™¾åˆ†æ¯”å¯¹ä¸»æˆåˆ†è¿›è¡Œæ’åã€‚åœ¨è¿™ä¸ªä¾‹å­ä¸­ï¼Œæˆ‘ä»¬å¯ä»¥è§‚å¯Ÿåˆ° PC9-10 å‘¨å›´çš„â€œè‚˜éƒ¨â€ï¼Œè¿™è¡¨æ˜å¤§éƒ¨åˆ†çœŸå®ä¿¡å·æ˜¯åœ¨å‰ 10 ä¸ª PC ä¸­æ•è·çš„ã€‚
```r
ElbowPlot(pbmc)
```
![](images/2021-10-24-22-18-34.png)

## æ•´åˆå¤šä¸ªscRNA-seqæ•°æ®é›†

----------------------------------------------------

#### ä¸seuratæ ‡å‡†å¤„ç†æµç¨‹ä¸€è‡´

<mark>ä¸¤ä¸ªæˆ–å¤šä¸ªå•ç»†èƒæ•°æ®é›†çš„è”åˆåˆ†æã€‚</mark>åœ¨æ ‡å‡†å·¥ä½œæµç¨‹ä¸‹ï¼Œè¯†åˆ«å­˜åœ¨äºå¤šä¸ªæ•°æ®é›†ä¸­çš„ç»†èƒç¾¤å¯èƒ½ä¼šå‡ºç°é—®é¢˜ã€‚Seurat v4 åŒ…æ‹¬ä¸€ç»„æ–¹æ³•æ¥åŒ¹é…ï¼ˆæˆ–â€œå¯¹é½â€ï¼‰è·¨æ•°æ®é›†çš„å…±äº«ç»†èƒç¾¤ã€‚è¿™äº›æ–¹æ³•é¦–å…ˆè¯†åˆ«å¤„äºåŒ¹é…ç”Ÿç‰©å­¦çŠ¶æ€ï¼ˆâ€œé”šâ€ï¼‰çš„è·¨æ•°æ®é›†ç»†èƒå¯¹ï¼Œæ—¢å¯ç”¨äºæ ¡æ­£æ•°æ®é›†ä¹‹é—´çš„æŠ€æœ¯å·®å¼‚ï¼ˆå³æ‰¹é‡æ•ˆåº”æ ¡æ­£ï¼‰ï¼Œä¹Ÿå¯ç”¨äºè¿›è¡Œæ¯”è¾ƒçš„ scRNA-seq åˆ†æè·¨å®éªŒæ¡ä»¶ã€‚

æˆ‘ä»¬å±•ç¤ºäº†[Stuart*ã€Butler* ç­‰äºº 2019 å¹´](https://www.cell.com/cell/fulltext/S0092-8674(19)30559-8)æ‰€è¿°çš„ scRNA-seq æ•´åˆæ–¹æ³•ï¼Œä»¥å¯¹å¤„äº[é™æ¯æˆ–å¹²æ‰°ç´ åˆºæ¿€çŠ¶æ€](https://www.nature.com/articles/nbt.4042)çš„äººç±»å…ç–«ç»†èƒ (PBMC) è¿›è¡Œæ¯”è¾ƒåˆ†æã€‚

```R
# Setup the Seurat objects
library(Seurat)
library(SeuratData)
library(patchwork)

# load dataset
LoadData("ifnb")

# load seurat data(a sample)
dir <- dir("data/")
dir <- paste0("data/", dir)
samples_name = c('ctrl', 'stim')

##ä½¿ç”¨å¾ªç¯å‘½ä»¤æ‰¹é‡åˆ›å»ºseuratå¯¹è±¡
scRNAlist <- list()
for(i in 1:length(dir)){
#Insufficient data values to produce 24 bins.
counts <- Read10X(data.dir = dir[i])
scRNAlist[[i]] <- Read10X(data.dir = dir[i]) %>%
                   CreateSeuratObject(project=samples_name[i],
                   #min.cells=3,#å¯ä»¥æŒ‡å®šæ¯ä¸ªæ ·æœ¬æœ€å°‘çš„ç»†èƒæ•°ç›®
                   min.features = 100)
#ç»™ç»†èƒbarcodeåŠ ä¸ªå‰ç¼€ï¼Œé˜²æ­¢åˆå¹¶åbarcodeé‡å
scRNAlist[[i]] <- RenameCells(scRNAlist[[i]], add.cell.id = samples_name[i])
}
#ç»™åˆ—è¡¨å‘½åå¹¶ä¿å­˜æ•°æ®
names(scRNAlist) <- samples_name

# æ£€æŸ¥åˆå¹¶çš„å¯¹è±¡æ˜¯å¦å…·æœ‰é€‚å½“çš„ç‰¹å®šäºæ ·æœ¬çš„å‰ç¼€
head(scRNAlist$ctrl@meta.data)
head(scRNAlist$stim@meta.data)

# split the dataset into a list of two seurat objects (stim and CTRL)
ifnb.list <- SplitObject(ifnb, split.by = "stim")

# normalize and identify variable features for each dataset independently
ifnb.list <- lapply(X = ifnb.list, FUN = function(x) {
    x <- NormalizeData(x)
    x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})

# select features that are repeatedly variable across datasets for integration
features <- SelectIntegrationFeatures(object.list = ifnb.list)

# Perform integration
immune.anchors <- FindIntegrationAnchors(object.list = ifnb.list, anchor.features = features)
# this command creates an 'integrated' data assay
immune.combined <- IntegrateData(anchorset = immune.anchors)

# Perform an integrated analysis
# specify that we will perform downstream analysis on the corrected data note that the
# original unmodified data still resides in the 'RNA' assay
DefaultAssay(immune.combined) <- "integrated"

# Run the standard workflow for visualization and clustering
immune.combined <- ScaleData(immune.combined, verbose = FALSE)
immune.combined <- RunPCA(immune.combined, npcs = 30, verbose = FALSE)
immune.combined <- RunUMAP(immune.combined, reduction = "pca", dims = 1:30)
immune.combined <- FindNeighbors(immune.combined, reduction = "pca", dims = 1:30)
immune.combined <- FindClusters(immune.combined, resolution = 0.5)
# Visualization
p1 <- DimPlot(immune.combined, reduction = "umap", group.by = "stim")
p2 <- DimPlot(immune.combined, reduction = "umap", label = TRUE, repel = TRUE)
p1 + p2
```

![](images/viz-1.png)



```R
DimPlot(immune.combined, reduction = "umap", split.by = "stim")
```

![](images/split.dim-1.png)

```R
# Identify conserved cell type markers
# For performing differential expression after integration, we switch back to the original
# data
DefaultAssay(immune.combined) <- "RNA"
nk.markers <- FindConservedMarkers(immune.combined, ident.1 = 6, grouping.var = "stim", verbose = FALSE)
head(nk.markers)

FeaturePlot(immune.combined, features = c("CD3D", "SELL", "CREM", "CD8A", "GNLY", "CD79A", "FCGR3A",
    "CCL2", "PPBP"), min.cutoff = "q9")
```

è¯¥æ­¥éª¤ä¸ºäº†<mark>è¯†åˆ«è·¨æ¡ä»¶ä¿å®ˆçš„å…¸å‹ç»†èƒç±»å‹æ ‡è®°åŸºå› </mark>mark>ï¼Œæˆ‘ä»¬æä¾›äº†è¯¥`FindConservedMarkers()`åŠŸèƒ½ã€‚æ­¤å‡½æ•°å¯¹æ¯ä¸ªæ•°æ®é›†/ç»„æ‰§è¡Œå·®å¼‚åŸºå› è¡¨è¾¾æµ‹è¯•ï¼Œå¹¶ä½¿ç”¨ MetaDE R åŒ…ä¸­çš„å…ƒåˆ†ææ–¹æ³•ç»„åˆ p å€¼ã€‚ä¾‹å¦‚ï¼Œæˆ‘ä»¬å¯ä»¥è®¡ç®—åœ¨ç°‡ 6ï¼ˆNK ç»†èƒï¼‰ä¸­æ— è®ºåˆºæ¿€æ¡ä»¶å¦‚ä½•éƒ½æ˜¯ä¿å®ˆæ ‡è®°çš„åŸºå› ã€‚

![](images/annotate-1.png)

```R
immune.combined <- RenameIdents(immune.combined, `0` = "CD14 Mono", `1` = "CD4 Naive T", `2` = "CD4 Memory T",
    `3` = "CD16 Mono", `4` = "B", `5` = "CD8 T", `6` = "NK", `7` = "T activated", `8` = "DC", `9` = "B Activated",
    `10` = "Mk", `11` = "pDC", `12` = "Eryth", `13` = "Mono/Mk Doublets", `14` = "HSPC")
DimPlot(immune.combined, label = TRUE)
```

![](images/annotate-2.png)

`DotPlot()`å¸¦æœ‰`split.by`å‚æ•°çš„å‡½æ•°å¯ç”¨äºæŸ¥çœ‹è·¨æ¡ä»¶çš„ä¿å®ˆç»†èƒç±»å‹æ ‡è®°ï¼Œæ˜¾ç¤ºè¡¨è¾¾æ°´å¹³å’Œè¡¨è¾¾ä»»ä½•ç»™å®šåŸºå› çš„ç°‡ä¸­ç»†èƒçš„ç™¾åˆ†æ¯”ã€‚åœ¨è¿™é‡Œï¼Œæˆ‘ä»¬ä¸º 14 ä¸ªèšç±»ä¸­çš„æ¯ä¸€ä¸ªç»˜åˆ¶äº† 2-3 ä¸ªå¼ºæ ‡è®°åŸºå› ã€‚

```R
Idents(immune.combined) <- factor(Idents(immune.combined), levels = c("HSPC", "Mono/Mk Doublets",
    "pDC", "Eryth", "Mk", "DC", "CD14 Mono", "CD16 Mono", "B Activated", "B", "CD8 T", "NK", "T activated",
    "CD4 Naive T", "CD4 Memory T"))
markers.to.plot <- c("CD3D", "CREM", "HSPH1", "SELL", "GIMAP5", "CACYBP", "GNLY", "NKG7", "CCL5",
    "CD8A", "MS4A1", "CD79A", "MIR155HG", "NME1", "FCGR3A", "VMO1", "CCL2", "S100A9", "HLA-DQA1",
    "GPR183", "PPBP", "GNG11", "HBA2", "HBB", "TSPAN13", "IL3RA", "IGJ", "PRSS57")
DotPlot(immune.combined, features = markers.to.plot, cols = c("blue", "red"), dot.scale = 8, split.by = "stim") +
    RotatedAxis()
```

![](images/splitdotplot-1.png)

* è·¨æ¡ä»¶è¯†åˆ«å·®å¼‚è¡¨è¾¾åŸºå› 

ç°åœ¨æˆ‘ä»¬å·²ç»å¯¹é½äº†å—åˆºæ¿€å’Œæ§åˆ¶ç»†èƒï¼Œæˆ‘ä»¬å¯ä»¥å¼€å§‹è¿›è¡Œæ¯”è¾ƒåˆ†æå¹¶æŸ¥çœ‹åˆºæ¿€å¼•èµ·çš„å·®å¼‚ã€‚å¹¿æ³›è§‚å¯Ÿè¿™äº›å˜åŒ–çš„ä¸€ç§æ–¹æ³•æ˜¯ç»˜åˆ¶å—åˆºæ¿€ç»†èƒå’Œå¯¹ç…§ç»†èƒçš„å¹³å‡è¡¨è¾¾ï¼Œå¹¶åœ¨æ•£ç‚¹å›¾ä¸Šå¯»æ‰¾è§†è§‰å¼‚å¸¸å€¼çš„åŸºå› ã€‚åœ¨è¿™é‡Œï¼Œæˆ‘ä»¬å–å—åˆºæ¿€å’Œå¯¹ç…§å¹¼ç¨š T ç»†èƒå’Œ CD14 å•æ ¸ç»†èƒç¾¤çš„å¹³å‡è¡¨è¾¾ï¼Œå¹¶ç”Ÿæˆæ•£ç‚¹å›¾ï¼Œçªå‡ºæ˜¾ç¤ºå¯¹å¹²æ‰°ç´ åˆºæ¿€æœ‰æ˜¾ç€ååº”çš„åŸºå› ã€‚

```R
library(ggplot2)
library(cowplot)
theme_set(theme_cowplot())
t.cells <- subset(immune.combined, idents = "CD4 Naive T")
Idents(t.cells) <- "stim"
avg.t.cells <- as.data.frame(log1p(AverageExpression(t.cells, verbose = FALSE)$RNA))
avg.t.cells$gene <- rownames(avg.t.cells)

cd14.mono <- subset(immune.combined, idents = "CD14 Mono")
Idents(cd14.mono) <- "stim"
avg.cd14.mono <- as.data.frame(log1p(AverageExpression(cd14.mono, verbose = FALSE)$RNA))
avg.cd14.mono$gene <- rownames(avg.cd14.mono)

genes.to.label = c("ISG15", "LY6E", "IFI6", "ISG20", "MX1", "IFIT2", "IFIT1", "CXCL10", "CCL8")
p1 <- ggplot(avg.t.cells, aes(CTRL, STIM)) + geom_point() + ggtitle("CD4 Naive T Cells")
p1 <- LabelPoints(plot = p1, points = genes.to.label, repel = TRUE)
p2 <- ggplot(avg.cd14.mono, aes(CTRL, STIM)) + geom_point() + ggtitle("CD14 Monocytes")
p2 <- LabelPoints(plot = p2, points = genes.to.label, repel = TRUE)
p1 + p2
```

![](images/scatterplots-1.png)

åœ¨ä¸åŒæ¡ä»¶ä¸‹è¯†åˆ«å‡ºå¸¸è§çš„ç»†èƒç±»å‹ï¼Œæ‰€ä»¥æˆ‘ä»¬å¯ä»¥<mark>è¯¢é—®ç›¸åŒç±»å‹ç»†èƒåœ¨ä¸åŒæ¡ä»¶ä¸‹å“ªäº›åŸºå› ä¼šå‘ç”Ÿå˜åŒ–</mark>ã€‚é¦–å…ˆï¼Œæˆ‘ä»¬åœ¨ meta.data æ§½ä¸­åˆ›å»ºä¸€ä¸ªåˆ—æ¥ä¿å­˜ç»†èƒç±»å‹å’Œåˆºæ¿€ä¿¡æ¯ï¼Œå¹¶å°†å½“å‰æ ‡è¯†åˆ‡æ¢åˆ°è¯¥åˆ—ã€‚ç„¶åæˆ‘ä»¬`FindMarkers()`ç”¨æ¥å¯»æ‰¾åˆºæ¿€å’Œå¯¹ç…§ B ç»†èƒä¹‹é—´ä¸åŒçš„åŸºå› ã€‚è¯·æ³¨æ„ï¼Œæ­¤å¤„æ˜¾ç¤ºçš„è®¸å¤šé¡¶çº§åŸºå› ä¸æˆ‘ä»¬ä¹‹å‰ç»˜åˆ¶çš„æ ¸å¿ƒå¹²æ‰°ç´ ååº”åŸºå› ç›¸åŒã€‚æ­¤å¤–ï¼Œæˆ‘ä»¬çœ‹åˆ°çš„ CXCL10 ç­‰ç‰¹å®šäºå•æ ¸ç»†èƒå’Œ B ç»†èƒå¹²æ‰°ç´ ååº”çš„åŸºå› åœ¨æ­¤åˆ—è¡¨ä¸­ä¹Ÿæ˜¾ç¤ºå‡ºéå¸¸é‡è¦çš„æ„ä¹‰ã€‚

```R
immune.combined$celltype.stim <- paste(Idents(immune.combined), immune.combined$stim, sep = "_")
immune.combined$celltype <- Idents(immune.combined)
Idents(immune.combined) <- "celltype.stim"
b.interferon.response <- FindMarkers(immune.combined, ident.1 = "B_STIM", ident.2 = "B_CTRL", verbose = FALSE)
head(b.interferon.response, n = 15)
```

å¯è§†åŒ–åœ¨åŸºå› è¡¨è¾¾çš„è¿™äº›å˜åŒ–çš„å¦ä¸€ç§æœ‰ç”¨çš„æ–¹æ³•æ˜¯ä½¿ç”¨`split.by`é€‰é¡¹ç»™`FeaturePlot()`æˆ–`VlnPlot()`åŠŸèƒ½ã€‚è¿™å°†æ˜¾ç¤ºç»™å®šåŸºå› åˆ—è¡¨çš„ç‰¹å¾å›¾ï¼ŒæŒ‰åˆ†ç»„å˜é‡ï¼ˆæ­¤å¤„ä¸ºåˆºæ¿€æ¡ä»¶ï¼‰æ‹†åˆ†ã€‚CD3D å’Œ GNLY ç­‰åŸºå› æ˜¯å…¸å‹çš„ç»†èƒç±»å‹æ ‡è®°ï¼ˆç”¨äº T ç»†èƒå’Œ NK/CD8 T ç»†èƒï¼‰ï¼Œå®ƒä»¬å‡ ä¹ä¸å—å¹²æ‰°ç´ åˆºæ¿€çš„å½±å“ï¼Œå¹¶ä¸”åœ¨å¯¹ç…§ç»„å’Œå—åˆºæ¿€ç»„ä¸­æ˜¾ç¤ºå‡ºç›¸ä¼¼çš„åŸºå› è¡¨è¾¾æ¨¡å¼ã€‚å¦ä¸€æ–¹é¢ï¼ŒIFI6 å’Œ ISG15 æ˜¯æ ¸å¿ƒå¹²æ‰°ç´ ååº”åŸºå› ï¼Œå¹¶åœ¨æ‰€æœ‰ç»†èƒç±»å‹ä¸­ç›¸åº”ä¸Šè°ƒã€‚æœ€åï¼ŒCD14 å’Œ CXCL10 æ˜¯æ˜¾ç¤ºç»†èƒç±»å‹ç‰¹å¼‚æ€§å¹²æ‰°ç´ ååº”çš„åŸºå› ã€‚åˆºæ¿€ CD14 å•æ ¸ç»†èƒå CD14 è¡¨è¾¾é™ä½ï¼Œè¿™å¯èƒ½å¯¼è‡´ç›‘ç£åˆ†ææ¡†æ¶ä¸­çš„é”™è¯¯åˆ†ç±»ï¼Œå¼ºè°ƒäº†ç»¼åˆåˆ†æçš„ä»·å€¼ã€‚

```R
FeaturePlot(immune.combined, features = c("CD3D", "GNLY", "IFI6"), split.by = "stim", max.cutoff = 3,
    cols = c("grey", "red"))

plots <- VlnPlot(immune.combined, features = c("LYZ", "ISG15", "CXCL10"), split.by = "stim", group.by = "celltype",
    pt.size = 0, combine = FALSE)
wrap_plots(plots = plots, ncol = 1)
```

![](images/splitvln-1.png)

#### ä½¿ç”¨ SCTransform è§„èŒƒåŒ–çš„æ•°æ®é›†æ‰§è¡Œæ•´åˆ

åœ¨[Hafemeister å’Œ Satijaï¼Œ2019 å¹´](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1)ï¼Œä»‹ç»äº†ä¸€ç§åŸºäºæ­£åˆ™åŒ–è´ŸäºŒé¡¹å¼å›å½’çš„æ”¹è¿›çš„ scRNA-seq æ ‡å‡†åŒ–æ–¹æ³•ã€‚è¯¥æ–¹æ³•è¢«å‘½åä¸ºâ€œsctransformâ€ï¼Œå¹¶é¿å…äº†æ ‡å‡†è§„èŒƒåŒ–å·¥ä½œæµç¨‹çš„ä¸€äº›ç¼ºé™·ï¼ŒåŒ…æ‹¬æ·»åŠ ä¼ªè®¡æ•°å’Œå¯¹æ•°è½¬æ¢ã€‚

æœ‰ä¸€äº›å…³é”®åŒºåˆ«ï¼š

- é€šè¿‡ å•ç‹¬è§„èŒƒåŒ–æ•°æ®é›†`SCTransform()`ï¼Œè€Œä¸æ˜¯`NormalizeData()`åœ¨é›†æˆä¹‹å‰
- æ­£å¦‚æˆ‘ä»¬åœ¨[SCTransform å°æ’å›¾ä¸­](https://satijalab.org/seurat/articles/sctransform_vignette.html)è¿›ä¸€æ­¥è®¨è®ºçš„[é‚£æ ·](https://satijalab.org/seurat/articles/sctransform_vignette.html)ï¼Œæˆ‘ä»¬é€šå¸¸ä½¿ç”¨ 3,000 ä¸ªæˆ–æ›´å¤šç‰¹å¾æ¥åˆ†æ sctransform çš„ä¸‹æ¸¸ã€‚
- `PrepSCTIntegration()`åœ¨è¯†åˆ«é”šç‚¹ä¹‹å‰è¿è¡Œå‡½æ•°
- è¿è¡Œ`FindIntegrationAnchors()`æ—¶`IntegrateData()`ï¼Œè®¾ç½®`normalization.method`å‚æ•°ä¸ºå€¼`SCT`ã€‚
- è¿è¡ŒåŸºäº sctransform çš„å·¥ä½œæµï¼ˆåŒ…æ‹¬é›†æˆï¼‰æ—¶ï¼Œè¯·å‹¿è¿è¡Œè¯¥`ScaleData()`å‡½æ•°

```R
LoadData("ifnb")
ifnb.list <- SplitObject(ifnb, split.by = "stim")
ifnb.list <- lapply(X = ifnb.list, FUN = SCTransform)
features <- SelectIntegrationFeatures(object.list = ifnb.list, nfeatures = 3000)
ifnb.list <- PrepSCTIntegration(object.list = ifnb.list, anchor.features = features)
### è¿è¡ŒFindIntegrationAnchors()æ—¶IntegrateData()ï¼Œè®¾ç½®normalization.methodå‚æ•°ä¸ºå€¼SCTã€‚
immune.anchors <- FindIntegrationAnchors(object.list = ifnb.list, normalization.method = "SCT",
    anchor.features = features)
immune.combined.sct <- IntegrateData(anchorset = immune.anchors, normalization.method = "SCT")
immune.combined.sct <- RunPCA(immune.combined.sct, verbose = FALSE)
immune.combined.sct <- RunUMAP(immune.combined.sct, reduction = "pca", dims = 1:30)
p1 <- DimPlot(immune.combined.sct, reduction = "umap", group.by = "stim")
p2 <- DimPlot(immune.combined.sct, reduction = "umap", group.by = "seurat_annotations", label = TRUE,
    repel = TRUE)
p1 + p2
```

![](images/immunesca.cca.sct.split.dims-1.png)



## å°† Seurat ä¸å¤šæ¨¡å¼æ•°æ®ä¸€èµ·ä½¿ç”¨

ä»åŒä¸€ä¸ªç»†èƒåŒæ—¶æµ‹é‡å¤šç§æ•°æ®ç±»å‹çš„èƒ½åŠ›ï¼Œç§°ä¸ºå¤šæ¨¡æ€åˆ†æã€‚é¦–å…ˆï¼ŒåŠ è½½ä¸¤ä¸ªè®¡æ•°çŸ©é˜µï¼šä¸€ä¸ªç”¨äº RNA æµ‹é‡ï¼Œä¸€ä¸ªç”¨äºæŠ—ä½“è¡ç”Ÿæ ‡ç­¾ (ADT)ã€‚

1. åŠ è½½æ•°æ®

```R
library(Seurat)
library(ggplot2)
library(patchwork)

# Load in the RNA UMI matrix

# Note that this dataset also contains ~5% of mouse cells, which we can use as negative
# controls for the protein measurements. For this reason, the gene expression matrix has
# HUMAN_ or MOUSE_ appended to the beginning of each gene.
cbmc.rna <- as.sparse(read.csv(file = "../data/GSE100866_CBMC_8K_13AB_10X-RNA_umi.csv.gz", sep = ",",
    header = TRUE, row.names = 1))

# To make life a bit easier going forward, we're going to discard all but the top 100 most
# highly expressed mouse genes, and remove the 'HUMAN_' from the CITE-seq prefix
cbmc.rna <- CollapseSpeciesExpressionMatrix(cbmc.rna)

# Load in the ADT UMI matrix
cbmc.adt <- as.sparse(read.csv(file = "../data/GSE100866_CBMC_8K_13AB_10X-ADT_umi.csv.gz", sep = ",",
    header = TRUE, row.names = 1))

# Note that since measurements were made in the same cells, the two matrices have identical column names
all.equal(colnames(cbmc.rna), colnames(cbmc.adt))
```

1. è®¾ç½® Seurat å¯¹è±¡ï¼Œæ·»åŠ  RNA å’Œè›‹ç™½è´¨æ•°æ®

```R
# creates a Seurat object based on the scRNA-seq data
cbmc <- CreateSeuratObject(counts = cbmc.rna)

# We can see that by default, the cbmc object contains an assay storing RNA measurement
Assays(cbmc)

# create a new assay to store ADT information
adt_assay <- CreateAssayObject(counts = cbmc.adt)

# add this assay to the previously created Seurat object
cbmc[["ADT"]] <- adt_assay

# Validate that the object now contains multiple assays
Assays(cbmc)

# Switch the default to ADT(switch)
DefaultAssay(cbmc) <- "ADT"
DefaultAssay(cbmc)
```

1. åŸºäºå®ƒä»¬çš„ scRNA-seq çš„æ•°æ®å¯¹ç»†èƒè¿›è¡Œèšç±»

```R
# Note that all operations below are performed on the RNA assay Set and verify that the
# default assay is RNA
DefaultAssay(cbmc) <- "RNA"
DefaultAssay(cbmc)

# perform visualization and clustering steps
cbmc <- NormalizeData(cbmc)
cbmc <- FindVariableFeatures(cbmc)
cbmc <- ScaleData(cbmc)
cbmc <- RunPCA(cbmc, verbose = FALSE)
cbmc <- FindNeighbors(cbmc, dims = 1:30)
cbmc <- FindClusters(cbmc, resolution = 0.8, verbose = FALSE)
cbmc <- RunUMAP(cbmc, dims = 1:30)
DimPlot(cbmc, label = TRUE)
```

![](images/cluster1-1.png)

1. å¹¶æ’æ˜¾ç¤ºå¤šç§æ¨¡å¼çš„ç»“æœ

```R
# Normalize ADT data,
DefaultAssay(cbmc) <- "ADT"
cbmc <- NormalizeData(cbmc, normalization.method = "CLR", margin = 2)
DefaultAssay(cbmc) <- "RNA"

# Note that the following command is an alternative but returns the same result
cbmc <- NormalizeData(cbmc, normalization.method = "CLR", margin = 2, assay = "ADT")

# Now, we will visualize CD14 levels for RNA and protein By setting the default assay, we can
# visualize one or the other
DefaultAssay(cbmc) <- "ADT"
p1 <- FeaturePlot(cbmc, "CD19", cols = c("lightgrey", "darkgreen")) + ggtitle("CD19 protein")
DefaultAssay(cbmc) <- "RNA"
p2 <- FeaturePlot(cbmc, "CD19") + ggtitle("CD19 RNA")

# place plots side-by-side
p1 | p2
```

![](images/vis-1.png)

```R
# Alternately, we can use specific assay keys to specify a specific modality Identify the key
# for the RNA and protein assays
Key(cbmc[["RNA"]])
Key(cbmc[["ADT"]])
# Now, we can include the key in the feature name, which overrides the default assay
p1 <- FeaturePlot(cbmc, "adt_CD19", cols = c("lightgrey", "darkgreen")) + ggtitle("CD19 protein")
p2 <- FeaturePlot(cbmc, "rna_CD19") + ggtitle("CD19 RNA")
p1 | p2
```

![](images/vis-2.png)

1. è¯†åˆ« scRNA-seq ç°‡çš„ç»†èƒè¡¨é¢æ ‡è®°

```R
# as we know that CD19 is a B cell marker, we can identify cluster 6 as expressing CD19 on the
# surface
VlnPlot(cbmc, "adt_CD19")
# we can also identify alternative protein and RNA markers for this cluster through
# differential expression
adt_markers <- FindMarkers(cbmc, ident.1 = 6, assay = "ADT")
rna_markers <- FindMarkers(cbmc, ident.1 = 6, assay = "RNA")
head(adt_markers)
```

![](images/markers-1.png)

```R
# we can also identify alternative protein and RNA markers for this cluster through
# differential expression
adt_markers <- FindMarkers(cbmc, ident.1 = 6, assay = "ADT")
rna_markers <- FindMarkers(cbmc, ident.1 = 6, assay = "RNA")

head(adt_markers)
```

1. å¤šæ¨¡å¼æ•°æ®çš„å…¶ä»–å¯è§†åŒ–

```R
# Draw ADT scatter plots (like biaxial plots for FACS). Note that you can even 'gate' cells if
# desired by using HoverLocator and FeatureLocator
FeatureScatter(cbmc, feature1 = "adt_CD19", feature2 = "adt_CD3")
```

![](images/viz.cite.two-1.png)

```R
# view relationship between protein and RNA
FeatureScatter(cbmc, feature1 = "adt_CD3", feature2 = "rna_CD3E")
```

![](images/viz.cite.two-2.png)

```R
FeatureScatter(cbmc, feature1 = "adt_CD4", feature2 = "adt_CD8")
```

![](images/viz.cite.two-3.png)

```R
# Let's look at the raw (non-normalized) ADT counts. You can see the values are quite high,
# particularly in comparison to RNA values. This is due to the significantly higher protein
# copy number in cells, which significantly reduces 'drop-out' in ADT data
FeatureScatter(cbmc, feature1 = "adt_CD4", feature2 = "adt_CD8", slot = "counts")
```

![](images/viz.cite.two-4.png)

1. ä» 10X å¤šæ¨¡æ€å®éªŒä¸­åŠ è½½æ•°æ®

Seurat è¿˜èƒ½å¤Ÿåˆ†ææ¥è‡ªä½¿ç”¨ CellRanger v3 å¤„ç†çš„å¤šæ¨¡å¼ 10X å®éªŒçš„æ•°æ®ï¼›ä¾‹å¦‚ï¼Œæˆ‘ä»¬ä½¿ç”¨ 7,900 ä¸ªå¤–å‘¨è¡€å•ä¸ªæ ¸ç»†èƒ (PBMC) çš„æ•°æ®é›†é‡æ–°åˆ›å»ºäº†ä¸Šé¢çš„å›¾ã€‚

```R
pbmc10k.data <- Read10X(data.dir = "../data/pbmc10k/filtered_feature_bc_matrix/")
rownames(x = pbmc10k.data[["Antibody Capture"]]) <- gsub(pattern = "_[control_]*TotalSeqB", replacement = "",
    x = rownames(x = pbmc10k.data[["Antibody Capture"]]))

pbmc10k <- CreateSeuratObject(counts = pbmc10k.data[["Gene Expression"]], min.cells = 3, min.features = 200)
pbmc10k <- NormalizeData(pbmc10k)
pbmc10k[["ADT"]] <- CreateAssayObject(pbmc10k.data[["Antibody Capture"]][, colnames(x = pbmc10k)])
pbmc10k <- NormalizeData(pbmc10k, assay = "ADT", normalization.method = "CLR")

plot1 <- FeatureScatter(pbmc10k, feature1 = "adt_CD19", feature2 = "adt_CD3", pt.size = 1)
plot2 <- FeatureScatter(pbmc10k, feature1 = "adt_CD4", feature2 = "adt_CD8a", pt.size = 1)
plot3 <- FeatureScatter(pbmc10k, feature1 = "adt_CD3", feature2 = "CD3E", pt.size = 1)
(plot1 + plot2 + plot3) & NoLegend()
```

![](images/pbmc10x-1.png)

## ä¸‹æ¸¸åˆ†æ

### ç»†èƒæ°´å¹³
#### ç°‡
##### èšç±»åˆ†æ
1. SeuratåŒ…

åº”ç”¨ï¼š scRNA-seq æ•°æ®ã€CyTOF æ•°æ®
* è¿™äº›æ–¹æ³•å°†ç»†èƒåµŒå…¥åˆ°ä¸€ä¸ªå›¾ç»“æ„ä¸­â€”â€”ä¾‹å¦‚ä¸€ä¸ª K-æœ€è¿‘é‚» (KNN) å›¾ï¼Œåœ¨å…·æœ‰ç›¸ä¼¼åŸºå› è¡¨è¾¾æ¨¡å¼çš„å•å…ƒä¹‹é—´ç»˜åˆ¶è¾¹ï¼Œç„¶åå°è¯•å°†è¯¥å›¾åˆ’åˆ†ä¸ºé«˜åº¦äº’è¿çš„â€œå‡†é›†å›¢â€æˆ–â€œç¾¤ä½“â€ã€‚
* é¦–å…ˆåŸºäº PCA ç©ºé—´ä¸­çš„æ¬§å‡ é‡Œå¾·è·ç¦»æ„å»ºä¸€ä¸ª KNN å›¾ï¼Œå¹¶åŸºäºå…¶å±€éƒ¨é‚»åŸŸä¸­çš„å…±äº«é‡å ï¼ˆJaccard ç›¸ä¼¼æ€§ï¼‰ç»†åŒ–ä»»æ„ä¸¤ä¸ªå•å…ƒä¹‹é—´çš„è¾¹æƒé‡ã€‚æ­¤æ­¥éª¤ä½¿ç”¨è¯¥`FindNeighbors()`å‡½æ•°æ‰§è¡Œï¼Œå¹¶å°†å…ˆå‰å®šä¹‰çš„æ•°æ®é›†ç»´åº¦ï¼ˆå‰ 10 ä¸ª PCï¼‰ä½œä¸ºè¾“å…¥ã€‚
* ä¸ºäº†å¯¹ç»†èƒè¿›è¡Œèšç±»ï¼Œæˆ‘ä»¬æ¥ä¸‹æ¥åº”ç”¨æ¨¡å—åŒ–ä¼˜åŒ–æŠ€æœ¯ï¼Œä¾‹å¦‚ Louvain ç®—æ³•ï¼ˆé»˜è®¤ï¼‰æˆ– SLM ï¼Œä»¥è¿­ä»£æ–¹å¼å°†ç»†èƒåˆ†ç»„åœ¨ä¸€èµ·ï¼Œç›®æ ‡æ˜¯ä¼˜åŒ–æ ‡å‡†æ¨¡å—åŒ–å‡½æ•°. è¯¥FindClusters()å‡½æ•°å®ç°äº†è¿™ä¸ªè¿‡ç¨‹ï¼Œå¹¶åŒ…å«ä¸€ä¸ªåˆ†è¾¨ç‡å‚æ•°ï¼Œç”¨äºè®¾ç½®ä¸‹æ¸¸èšç±»çš„â€œåˆ†è¾¨ç‡ï¼ˆresolutionï¼‰â€ï¼Œ<mark>åˆ†è¾¨ç‡å¢åŠ ä¼šå¾—åˆ°æ›´å¤šçš„ç°‡ã€‚æˆ‘ä»¬å‘ç°ï¼Œå°†è¿™ä¸ªå‚æ•°è®¾ç½®åœ¨ 0.4-1.2 ä¹‹é—´é€šå¸¸ä¼šä¸ºå¤§çº¦ 3K ç»†èƒçš„å•ç»†èƒæ•°æ®é›†è¿”å›è¾ƒå¥½çš„èšç±»ç»“æœã€‚</mark>å¯¹äºè¾ƒå¤§çš„æ•°æ®é›†ï¼Œæœ€ä½³åˆ†è¾¨ç‡é€šå¸¸ä¼šå¢åŠ ã€‚å¯ä»¥ä½¿ç”¨è¯¥`Idents()`å‡½æ•°æ‰¾åˆ°ç°‡ã€‚
```r
pbmc <- FindNeighbors(pbmc, dims = 1:10)
pbmc <- FindClusters(pbmc, resolution = 0.5)
```
![](images/2021-10-24-22-19-43.png)
```r
# Look at cluster IDs of the first 5 cells
head(Idents(pbmc), 5)
```
![](images/2021-10-24-22-20-14.png)

2. é€šè¿‡ä¼ªæ—¶é—´è¡¨è¾¾æ¨¡å¼èšç±»åŸºå› 

åœ¨ç ”ç©¶æ—¶é—´åºåˆ—åŸºå› è¡¨è¾¾ç ”ç©¶æ—¶å‡ºç°çš„ä¸€ä¸ªå¸¸è§é—®é¢˜æ˜¯ï¼šâ€œå“ªäº›åŸºå› éµå¾ªç›¸ä¼¼çš„åŠ¨åŠ›å­¦è¶‹åŠ¿â€ï¼ŸMonocle å¯ä»¥é€šè¿‡å¯¹å…·æœ‰ç›¸ä¼¼è¶‹åŠ¿çš„åŸºå› è¿›è¡Œåˆ†ç»„æ¥å¸®åŠ©æ‚¨å›ç­”è¿™ä¸ªé—®é¢˜ï¼Œå› æ­¤æ‚¨å¯ä»¥åˆ†æè¿™äº›ç»„ä»¥æŸ¥çœ‹å®ƒä»¬çš„å…±åŒç‚¹ã€‚Monocle æä¾›äº†ä¸€ç§å¯è§†åŒ–æ‰€æœ‰ä¼ªæ—¶é—´ä¾èµ–æ€§åŸºå› çš„ä¾¿æ·æ–¹æ³•ã€‚è¯¥å‡½æ•°`plot_pseudotime_heatmap`æ¥å—ä¸€ä¸ª`CellDataSet`å¯¹è±¡ï¼ˆé€šå¸¸åªåŒ…å«é‡è¦åŸºå› çš„ä¸€ä¸ªå­é›†ï¼‰å¹¶ç”Ÿæˆå¹³æ»‘çš„è¡¨è¾¾æ›²çº¿ï¼Œå°±åƒ`plot_genes_in_pseudotime`ã€‚ç„¶åï¼Œå®ƒå¯¹è¿™äº›åŸºå› è¿›è¡Œèšç±»å¹¶ä½¿ç”¨pheatmapåŒ…ç»˜åˆ¶å®ƒä»¬ï¼Œå¯ä»¥**å¯è§†åŒ–è·¨ä¼ªæ—¶é—´å…±å˜çš„åŸºå› æ¨¡å—**ã€‚
```r
diff_test_res <- differentialGeneTest(HSMM_myo[marker_genes,],
              fullModelFormulaStr = "~sm.ns(Pseudotime)")
sig_gene_names <- row.names(subset(diff_test_res, qval < 0.1))
plot_pseudotime_heatmap(HSMM_myo[sig_gene_names,],
                num_clusters = 3,
                cores = 1,
                show_rownames = T)
```
![](images/2021-10-25-14-30-50.png)







##### å¯¹ç°‡è¿›è¡Œç»†èƒç±»å‹æ³¨é‡Š
1. SeuratåŒ…
å°†æ— åèšç±»ä¸å·²çŸ¥ç»†èƒç±»å‹åŒ¹é…ï¼Œå¹¶è¿›è¡Œæ³¨é‡Šï¼Œ<mark>å±äºå…ˆéªŒçŸ¥è¯†è¿›è¡Œæ³¨é‡Šï¼Œä¸ºæ— ç›‘ç£å­¦ä¹ ã€‚</mark>
```r
new.cluster.ids <- c("Naive CD4 T", "CD14+ Mono", "Memory CD4 T", "B", "CD8 T", "FCGR3A+ Mono","NK", "DC", "Platelet")
names(new.cluster.ids) <- levels(pbmc)
pbmc <- RenameIdents(pbmc, new.cluster.ids)
DimPlot(pbmc, reduction = "umap", label = TRUE, pt.size = 0.5)+NoLegend()
saveRDS(pbmc, file = "../output/pbmc3k_final.rds")
```
![](images/2021-10-25-14-38-05.png)
2. SingleRåŒ…
SingleRåˆ©ç”¨çº¯ç»†èƒç±»å‹çš„å‚è€ƒè½¬å½•ç»„æ•°æ®é›†æ¥ç‹¬ç«‹æ¨æ–­æ¯ä¸ªå•ç»†èƒçš„ç»†èƒå¯èƒ½ç±»å‹ã€‚SingleRçš„æ³¨é‡Šä¸Seurat(ä¸€ä¸ªä¸ºscRNA-seqè®¾è®¡çš„å¤„ç†å’Œåˆ†æåŒ…)ç›¸ç»“åˆï¼Œä¸ºç ”ç©¶scRNA-seqæ•°æ®æä¾›äº†ä¸€ä¸ªå¼ºå¤§çš„å·¥å…·ã€‚SingleRæä¾›äº†å†…ç½®çš„åŒ…è£…å‡½æ•°ï¼Œå¯ä»¥ç”¨ä¸€ä¸ªå‡½æ•°è¿è¡Œå®Œæ•´çš„æµç¨‹ã€‚SingleRä¸Seuratäº¤äº’è¾ƒå¥½ï¼Œä½†æ˜¯ä¹Ÿå¯ä»¥ä½¿ç”¨ä»»ä½•å…¶ä»–scRNA-seqåŒ…ï¼Œä½¿ç”¨è¯¥åŒ…æ‰€æœ‰çš„æ•°æ®é›†,ç½‘å€å¦‚ä¸‹ã€‚
<https://bioconductor.org/packages/release/data/experiment/vignettes/scRNAseq/inst/doc/scRNAseq.html#references>

åŸç†ï¼šé¦–å…ˆï¼Œè®¡ç®—å‚è€ƒæ•°æ®é›†ä¸­æ¯ä¸ªæ ·æœ¬çš„å•ç»†èƒè¡¨è¾¾çš„spearmanç³»æ•°ã€‚ç›¸å…³åˆ†æä»…å¯¹å‚è€ƒæ•°æ®é›†ä¸­çš„å¯å˜åŸºå› ï¼ˆvariable genesï¼‰è¿›è¡Œã€‚æ¥ç€ï¼Œæ ¹æ®å‚è€ƒæ•°æ®é›†çš„å‘½åæ³¨é‡Šèšåˆæ¯ä¸ªç»†èƒç±»å‹çš„å¤šä¸ªç›¸å…³ç³»æ•°ï¼Œä»è€Œä¸ºæ¯ä¸ªç»†èƒç±»å‹æä¾›ä¸€ä¸ªå€¼ã€‚æ¥ç€SingleRå°†é‡æ–°è¿è¡Œç›¸å…³åˆ†æï¼Œä½†åªé’ˆå¯¹ä¸Šä¸€æ­¥ä¸­çš„ç›¸å…³æ€§è¾ƒé«˜çš„ç»†èƒç±»å‹ã€‚è¯¥åˆ†æä»…å¯¹è¿™äº›ç»†èƒç±»å‹ä¹‹é—´çš„å¯å˜åŸºå› è¿›è¡Œã€‚ç§»é™¤æœ€ä½å€¼çš„ç»†èƒç±»å‹(æˆ–æ¯”æœ€é«˜å€¼ä½0.05çš„è¾¹ç¼˜)ï¼Œç„¶åé‡å¤æ­¤æ­¥éª¤ï¼Œç›´åˆ°åªä¿ç•™ä¸¤ç§ç»†èƒç±»å‹ã€‚æœ€åä¸€æ¬¡è¿è¡Œåï¼Œä¸é¡¶éƒ¨å€¼å¯¹åº”çš„ç»†èƒç±»å‹è¢«åˆ†é…ç»™å•ä¸ªç»†èƒã€‚
```r
# æ ¹æ®éœ€æ±‚é€‰æ‹©ä¸Šè¿°ç½‘ç«™çš„æ•°æ®é›†ä½œä¸ºSingleRçš„å‚è€ƒæ•°æ®é›†
imm_ref <- DatabaseImmuneCellExpressionData()
# è¿›è¡Œç»†èƒç±»å‹æ³¨é‡Šï¼ˆä¸€èˆ¬é€‰æ‹©ç¬¬äºŒç§ï¼Œå¯ä»¥ç›´æ¥æ³¨é‡ŠSeuratèšç±»ç»“æœï¼‰
pred_single <- SingleR(test = pbmc@assays$SCT@data, 
                       ref = imm_ref, 
                       labels = imm_ref$label.main,
                       genes = "de",
                       de.method = "wilcox")

pred_cluster <- SingleR(test = pbmc@assays$SCT@data, 
                        ref = imm_ref, 
                        labels = imm_ref$label.main,
                        clusters = pbmc@meta.data$SCT_snn_res.0.6,
                        genes = "de",
                        de.method = "wilcox")

save(pred_single, file = "pred_single.rda")
save(pred_cluster, file = "pred_cluster.rda")

#
# plot
# 
x <- data.frame(seurat_celltype = Idents(pbmc),
                sr_single_first_ct = pred_single$first.labels,
                sr_single_ct = pred_single$labels,
                sr_single_pruned_ct = pred_single$pruned.labels,
                sr_cluster_first_ct = RenameCluster(pbmc@meta.data$SCT_snn_res.0.6, pred_cluster$first.labels),
                sr_cluster_ct = RenameCluster(pbmc@meta.data$SCT_snn_res.0.6, pred_cluster$labels),
                sr_cluster_pruned_ct = RenameCluster(pbmc@meta.data$SCT_snn_res.0.6, pred_cluster$pruned.labels),
                stringsAsFactors = TRUE)
pbmc@meta.data <- cbind(pbmc@meta.data, x, stringsAsFactors = TRUE)

plist = lapply(colnames(x), function(col){
  DimPlot(pbmc, reduction="umap", group.by = col, label = FALSE, pt.size = 1.5) +
    labs(title = col) + 
    scale_colour_d3("category20") + 
    rj.ftheme
})
ggsave("singer_compare_all.png", CombinePlots(plist, ncol = 3), width = 20, height = 15)

plist = lapply(colnames(x)[c(1,6,3)], function(col){
  DimPlot(pbmc, reduction="umap", group.by = col, label = FALSE, pt.size = 1.5) + labs(title = col) + scale_colour_d3("category20") + rj.ftheme
})
ggsave("singer_compare_ct.png", CombinePlots(plist, ncol = 3), width = 20, height = 5)


https://nbisweden.github.io/excelerate-scRNAseq/session-celltypeid/celltypeid.html

```


_______________

Updateï¼šSingleRåŒ…è¿›è¡Œæ”¹ç‰ˆå’Œæ›´æ–° å»ºè®®ä½¿ç”¨ä»¥ä¸‹çš„ä»£ç ã€‚

```R
# Cell_typing
# SingleRç»†èƒç±»å‹æ³¨é‡Š
  Seurat.data <- readRDS("Markers/Markers_0.5.rds")
# SCTransform()å‡½æ•°
  Seurat.data <- SCTransform(Seurat.data, vars.to.regress = "percent.mt")
# Load the SingleRdata  
  Seurat.data_for_SingleR <- GetAssayData(Seurat.data, slot="data")
  clusters=Seurat.data@meta.data$seurat_clusters
# æ ¹æ®éœ€æ±‚é€‰æ‹©ä¸Šè¿°ç½‘ç«™çš„æ•°æ®é›†ä½œä¸ºSingleRçš„å‚è€ƒæ•°æ®é›†
  library(SingleR)
  library(scRNAseq)
  library(celldex)
#### Human being ####
#####################
## Using multiple references
  library(celldex)
  hpca <- HumanPrimaryCellAtlasData(ensembl=TRUE)
  bpe <- BlueprintEncodeData(ensembl=TRUE)

## annotation
  pred.human <- SingleR(test = Seurat.data_for_SingleR, assay.type.test=1,
                      ref = list(BPE=bpe, HPCA=hpca), 
                      labels = list(bpe$label.main, hpca$label.main))
  
## Check the final label from the combined assignment.
  table(pred.human$labels) 
  head(pred.human$orig.results$BPE$labels)
  head(pred.human$orig.results$HPCA$labels)

## visualization
  Seurat.data$hpca_type <- pred.human$orig.results$HPCA$labels
  Seurat.data$bpe_typ <- pred.human$orig.results$BPE$labels
  
## visualization
  library(ggsci)
  p1 <- DimPlot(Seurat.data, group.by='hpca_type',pt.size = 3)+scale_color_ucscgb()
  p2 <- DimPlot(Seurat.data, group.by='bpe_type',pt.size = 3)+scale_color_ucscgb()
  library(cowplot)
  p<-plot_grid(p1, p2, nrow=1, ncol=2, labels=panel.labels)
  ggsave("singler.pdf", plot = p, width = 15, height = 10) 
  
  
####### Mouse #######
#####################
  library(celldex)
  #immgen <- ImmGenData()
  immgen <- celldex::ImmGenData()
  mouseRNA <- celldex::MouseRNAseqData()
  monacoImm <- celldex::MonacoImmuneData()

## annotation
  pred.mouse <- SingleR(test = Seurat.data_for_SingleR, assay.type.test=1,
                        ref = list(Immgen=immgen, MouseRNA=mouseRNA,MonacoImmune=monacoImm), 
                        labels = list(immgen$label.main, mouseRNA$label.main,monacoImm$label.main))
  
## Check the final label from the combined assignment.
  table(pred.mouse$labels) 
  
## Check the 'winning' reference for each cell.
  table(pred.mouse$reference)   
  head(pred.mouse$orig.results$Immgen$labels)  
  head(pred.mouse$orig.results$MouseRNA$labels)
  head(pred.mouse$orig.results$MonacoImmune$labels)
    
## visualization
  Seurat.data$immgen <- pred.mouse$orig.results$Immgen$labels
  Seurat.data$MouseRNA <- pred.mouse$orig.results$MouseRNA$labels
  Seurat.data$MonacoImmun <- pred.mouse$orig.results$MonacoImmune$labels
  
  library(ggsci)
  p1 <- DimPlot(Seurat.data, group.by='immgen',pt.size = 3)+scale_color_ucscgb()
  p2 <- DimPlot(Seurat.data, group.by='MouseRNA',pt.size = 3)+scale_color_ucscgb()
  p3 <- DimPlot(Seurat.data, group.by='MonacoImmun',pt.size = 3)+scale_color_ucscgb()
  library(cowplot)
  p<-plot_grid(p1, p2,p3,nrow=1, ncol=3)
  ggsave("singler.pdf", plot = p, width = 15, height = 10)   
```

3. GarnettåŒ…
    ï¼ˆåŸºäºç›‘ç£å­¦ä¹ çš„ä»å•ç»†èƒè¡¨è¾¾æ•°æ®ä¸­å®ç°è‡ªåŠ¨ç»†èƒç±»å‹åˆ†ç±»çš„è½¯ä»¶åŒ…ï¼‰

<font color=#FF0000>ï¼ï¼ï¼

è¯¥åŒ…ä¸å†æ”¯æŒåé¢æåˆ°çš„monocle2åŒ…ï¼Œç›®å‰åªæ”¯æŒmonocle3ã€‚</font>

åŸç†ï¼šè·å–å•ç»†èƒæ•°æ®å’Œç»†èƒç±»å‹å®šä¹‰(marker)æ–‡ä»¶ï¼Œå¹¶è®­ç»ƒä¸€ä¸ªåŸºäºå›å½’çš„åˆ†ç±»å™¨ã€‚ä¸€æ—¦è¢«è®­ç»ƒæˆä¸€ä¸ªé’ˆå¯¹æŸä¸€ç»„ç»‡/æ ·æœ¬ç±»å‹çš„ä¸€ä¸ªåˆ†ç±»å™¨ï¼Œå®ƒå°±å¯ä»¥åº”ç”¨äºä»ç›¸ä¼¼ç»„ç»‡ä¸­å¯¹æœªæ¥çš„æ•°æ®é›†è¿›è¡Œåˆ†ç±»ï¼Œå…¶ä¸­åˆ†ç±»å™¨å¯ä»¥æ˜¯ä¸‹è½½ç°æœ‰çš„åˆ†ç±»å™¨æˆ–è€…è®­ç»ƒè‡ªå·±çš„åˆ†ç±»å™¨ã€‚
```r
library(ggplot2)
library(ggsci)
library(Seurat)
library(dplyr)
library(tibble)
library(garnett)
library(monocle)
#
# 1. create monocle object from seurat
#
load("~/project/sc_standard_procedure/1_framework/Seurat/3.2.3_with_SCTransform/regress_nothing/pc10/pbmc.rda")
mat <- pbmc@assays$SCT@counts 
fdata <- data.frame(row.names = rownames(mat),
                    gene_short_name = rownames(mat),
                    stringsAsFactors = FALSE)
pdata <- data.frame(pbmc@meta.data,
                    pbmc@reductions$tsne@cell.embeddings,
                    stringsAsFactors = FALSE)
               
rownames(mat) <- rownames(fdata)
colnames(mat) <- rownames(pdata)

pd <- new("AnnotatedDataFrame", data = pdata)
fd <- new("AnnotatedDataFrame", data = fdata)
pbmc3k_cds <- newCellDataSet(as(mat, "dgCMatrix"),
                           phenoData = pd,
                           featureData = fd)
pbmc3k_cds <- estimateSizeFactors(pbmc3k_cds) # å½’ä¸€åŒ–

#
# 2. choose classifier and  classify
# 
classifier <- readRDS("~/project/sc_standard_procedure/2_annotation/garnett/pre_trained_classifier/hsPBMC_20191017.RDS")

library(org.Hs.eg.db)
pbmc3k_cds <- classify_cells(pbmc3k_cds, classifier,
                           db = org.Hs.eg.db,
                           cluster_extend = TRUE,
                           cds_gene_id_type = "SYMBOL")

table(pData(pbmc3k_cds)$cell_type)
table(pData(pbmc3k_cds)$cluster_ext_type)

#
# 3. plot
# 
x <- data.frame(seurat_celltype = Idents(pbmc),
                garnett_cell_type = pData(pbmc3k_cds)$cell_type,
                garnett_cluster_ext_type = pData(pbmc3k_cds)$cluster_ext_type,
                stringsAsFactors = TRUE)
pbmc@meta.data <- cbind(pbmc@meta.data, x, stringsAsFactors = TRUE)

plist = lapply(colnames(x), function(col){
  DimPlot(pbmc, reduction="umap", group.by = col, label = FALSE, pt.size = 1.5) +
    labs(title = col) + 
    scale_colour_d3("category20") + 
    rj.ftheme
})

ggsave("garnett_compare_all_1.png", CombinePlots(plist, ncol = 3), width = 16, height = 4)
ggsave("garnett_compare_all_2.png", CombinePlots(plist, ncol = 3), width = 16, height = 4)

writeLines(capture.output(sessionInfo()), "sessionInfo.txt")
#----------------------------------------------------------------
#                                                               -
#                             test                              -
#                                                               -
#----------------------------------------------------------------

#
# 1 create monocle object
#
mat <- Matrix::readMM(system.file("extdata", "exprs_sparse.mtx", package = "garnett"))

fdata <- read.table(system.file("extdata", "fdata.txt", package = "garnett"))
pdata <- read.table(system.file("extdata", "pdata.txt", package = "garnett"), sep="\t")
row.names(mat) <- row.names(fdata)
colnames(mat) <- row.names(pdata)

pd <- new("AnnotatedDataFrame", data = pdata)
fd <- new("AnnotatedDataFrame", data = fdata)
pbmc_cds <- newCellDataSet(as(mat, "dgCMatrix"),
                           phenoData = pd,
                           featureData = fd)

pbmc_cds <- estimateSizeFactors(pbmc_cds)

#
# 2 check marker
#
library(org.Hs.eg.db)
marker_file_path <- system.file("extdata", "pbmc_bad_markers.txt",
                                package = "garnett")
marker_check <- check_markers(pbmc_cds, marker_file_path,
                              db=org.Hs.eg.db,
                              cds_gene_id_type = "SYMBOL",
                              marker_file_gene_id_type = "SYMBOL")

plot_markers(marker_check)

#
# 3 train classifier
#
library(org.Hs.eg.db)
set.seed(260)

marker_file_path <- system.file("extdata", "pbmc_test.txt",
                                package = "garnett")
pbmc_classifier <- train_cell_classifier(cds = pbmc_cds,
                                         marker_file = marker_file_path,
                                         db=org.Hs.eg.db,
                                         cds_gene_id_type = "SYMBOL",
                                         num_unknown = 50,
                                   marker_file_gene_id_type = "SYMBOL")

feature_genes <- get_feature_genes(pbmc_classifier,
                                   node = "root",
                                   db = org.Hs.eg.db)

get_classifier_references(pbmc_classifier)

#
# 4 classify your cells
#
library(org.Hs.eg.db)
pbmc_cds <- classify_cells(pbmc_cds, pbmc_classifier,
                           db = org.Hs.eg.db,
                           cluster_extend = TRUE,
                           cds_gene_id_type = "SYMBOL")

table(pData(pbmc_cds)$cell_type)
table(pData(pbmc_cds)$cluster_ext_type)

library(ggplot2)
pdf("result.pdf")
qplot(tsne_1, tsne_2, color = cell_type, data = pData(pbmc_cds)) + theme_bw()
qplot(tsne_1, tsne_2, color = cluster_ext_type, data = pData(pbmc_cds)) + theme_bw()
qplot(tsne_1, tsne_2, color = FACS_type, data = pData(pbmc_cds)) + theme_bw()
dev.off()
```
4. scibetåŒ…

åŸç†ï¼šç»è¿‡å¤§é‡çš„ç»Ÿè®¡åˆ†æå’Œåç»­çš„å®éªŒéªŒè¯ï¼Œå•ç»†èƒåŸºå› è¡¨è¾¾çš„count(æ¯”å¦‚UMI count)çš„åˆ†å¸ƒå¯ä»¥ç”¨è´ŸäºŒé¡¹åˆ†å¸ƒå¾ˆå¥½çš„æ‹Ÿåˆ,ä¸”ç›¸åŒç»†èƒç±»å‹çš„å•ç»†èƒè¡¨è¾¾è°±æœä»åŒä¸€ä¸ªåˆ†å¸ƒã€‚å› æ­¤å°†å•ç»†èƒåŸºå› è¡¨è¾¾çš„countè¡¨ç¤ºä¸ºæ³Šæ¾åˆ†å¸ƒçš„ä¼½é©¬åˆ†å¸ƒçš„æ··åˆåˆ†å¸ƒï¼Œæ¥ç€æ ¹æ®è¡¨è¾¾æ•°æ®è®¡ç®—ä¿¡æ¯ç†µï¼ŒæŒ‘å‡ºé‚£äº›èƒ½è¡¨ç¤ºä¸åŒç°‡ç»†èƒä¹‹é—´å·®å¼‚è¡¨è¾¾çš„åŸºå› ï¼Œä½¿ç”¨E-testä»¥æœ‰ç›‘ç£å’Œå‚æ•°åŒ–çš„æ–¹å¼ä»è®­ç»ƒé›†ä¸­é€‰æ‹©ç»†èƒç±»å‹ç‰¹å¼‚æ€§åŸºå› ï¼Œä»¥å»é™¤å˜ˆæ‚çš„åŸºå› ä»¥åŠé€šè¿‡å‹ç¼©æ¨¡å‹æ¥åŠ é€Ÿä¸‹æ¸¸åˆ†ç±»ï¼Œæœ€åæ„å»ºå‡ºåŸºäºç›‘ç£å­¦ä¹ çš„æ¨¡å‹ï¼Œä»è€Œå¯¹ç»†èƒç±»å‹è¿›è¡Œæ³¨é‡Šã€‚
```r
library(scibet)
library(celldex)
library(ggplot2)
library(ggsci)
library(Seurat)
library(dplyr)
library(tibble)
library(viridis)
#----------------------------------------------------------------
#                                                               -
#                             e-test                            -
#                                                               -
#----------------------------------------------------------------

load("~/project/sc_standard_procedure/1_framework/Seurat/3.2.3_with_SCTransform/regress_nothing/pc10/pbmc.rda")

expr <- as.data.frame(t(pbmc@assays$SCT@data))
expr$label <- Idents(pbmc)

etest_gene <- SelectGene(expr, k = 500, r = FALSE)
ggsave("marker_heatmap.png", Marker_heatmap(expr, etest_gene), width = 30, height = 6)

#----------------------------------------------------------------
#                                                               -
#                         training model                        -
#                                                               -
#----------------------------------------------------------------


num2 = 0
num1 = 0.5
while ( (num2/num1) < 0.92 ) {
  
  # traning 
  tibble(
    ID = 1:nrow(expr),
    label = expr$label
  ) %>%
    dplyr::sample_frac(0.7) %>%
    dplyr::pull(ID) -> ID
  
  train_set <- expr[ID,]      #construct reference set
  test_set <- expr[-ID,]      #construct query set
  
  prd <- SciBet(train_set, test_set[,-ncol(test_set)], k = 500)
  
  # check accuracy
  ori_label = test_set$label
  label = prd
  num1 <- length(ori_label)
  num2 <- tibble(
    ori = ori_label,
    prd = label
  ) %>%
    dplyr::filter(ori == prd) %>%
    nrow(.)
  
  print(num2/num1)
}

model = Learn(train_set)
export_model = ExportModel(model)
write.csv(t(export_model), file = "pbmc3k_export_model.csv")

# check model
ggsave("confusion_heatmap.png", Confusion_heatmap(as.character(test_set$label), prd), width = 8, height = 7)

#----------------------------------------------------------------
#                                                               -
#                           annotation                          -
#                                                               -
#----------------------------------------------------------------

load("~/project/sc_standard_procedure/1_preprocessing_clustering/Seurat/3.2.3_with_SCTransform/regress_nothing/pc10/pbmc.rda")

expr <- as.data.frame(t(pbmc@assays$SCT@data))
ori_label <- pbmc@meta.data$SCT_snn_res.0.6

# reference
ref_list <- c("./reference/major_human_cell_types.csv",
              "./reference/GSE84465_scibet_core.csv",
              "./reference/GSE109774_scibet_core.csv",
              "./50gene_training_model/50gene_pbmc3k_export_model.csv",
              "./500gene_training_model/500gene_pbmc3k_export_model.csv")

# predict
label <- sapply(ref_list, function(ref){
  
  # load traning model
  model = read.csv(ref)
  colnames(model) = toupper(colnames(model))
  model = pro.core(model)
  
  # predict
  prd = LoadModel(model)
  prd(expr)
  
})

label_df <- data.frame(label)
colnames(label_df) <- stringr::str_split_fixed(colnames(label_df),"\\.",5)[,4]

# plot
x <- data.frame(seurat_celltype = Idents(pbmc),
                label_df,
                stringsAsFactors = TRUE)
pbmc@meta.data <- cbind(pbmc@meta.data, x, stringsAsFactors = TRUE)

plist = lapply(colnames(x), function(col){
  DimPlot(pbmc, reduction="umap", group.by = col, label = FALSE, pt.size = 1.5) +
    labs(title = col) + 
    #scale_colour_d3("category20") + 
    rj.ftheme
})
ggsave("scibet_compare_all_1.png", CombinePlots(plist[-4], ncol = 2), width = 11, height = 11)
ggsave("scibet_compare_all_2.png", CombinePlots(plist[4], ncol = 3), width = 11, height = 5)

writeLines(capture.output(sessionInfo()), "sessionInfo.txt")
```

5. PanglaoDB

[PanglaoDB]: https://panglaodb.se/index.html

![image-20220107171424266](/Users/hecate/Library/Application Support/typora-user-images/image-20220107171424266.png)

æ ¹æ®æ•°æ®ç±»å‹ï¼Œé€‰æ‹©æ‰€æƒ³è¦çš„markerå‚è€ƒåŸºå› é›†ã€‚

```r
celltype <- read.table('PanglaoDB_markers_27_Mar_2020.tsv',
                       sep = '\t',header = T,stringsAsFactors = F)

iHyperGD <- function(backGround,
                     geneNum.pathway,
                     geneNum.check,
                     geneNum.located){
  1-phyper(geneNum.located-1,
           geneNum.pathway,
           backGround - geneNum.pathway,
           geneNum.check,
           lower.tail = T)
}

hyperType <- function(genes,type,backGround){
  geneNum.pathway <- length(which(celltype$cell.type == type))
  geneNum.check   <- length(genes)
  geneNum.located <- sum(genes %in% celltype$official.gene.symbol[which(celltype$cell.type == type)])
  
  iHyperGD(backGround,geneNum.pathway,geneNum.check,geneNum.located)
}

cellTypeEst_hyper <- function(cluster,m,backGround = 2000,type = NA){
  genes <- m$gene[which(m$cluster == cluster)]
  
  p.val <- c()
  if(is.na(type)){
    for (Type in unique(celltype$cell.type)) {
      p.val <- c(p.val,hyperType(genes,Type,backGround))
    }
  }
  names(p.val) <- unique(celltype$cell.type)
  sort(p.val,decreasing = F)
}
# ä¸€æ¬¡ä¸€ä¸ªåˆ†ç°‡
cellTypeEst_hyper(2,m=pbmc.markers) 
for (i in (1:6)){
  print(cellTypeEst_hyper(i,m=pbmc.markers)[1])
}
```



#### æ„å»ºå•ç»†èƒè½¨è¿¹

ä¸»è¦ä½¿ç”¨Monocle2å’ŒMonocle3åŒ…ã€‚
> ä¼ªæ—¶é—´åˆ†æ

å…¶åŠŸèƒ½åŸºäºå•ç»†èƒè½¬å½•ç»„çš„è¡¨è¾¾çŸ©é˜µï¼Œé€šè¿‡æ— ç›‘ç£å­¦ä¹ ï¼ˆReversed Graph Embeddingç®—æ³•ï¼‰çš„æ–¹å¼å°†ç»†èƒç½®äºå‘è‚²è½¨è¿¹çš„ä¸åŒåˆ†æ”¯ä¸Šï¼Œä»è€Œæ¨¡æ‹Ÿç»†èƒç¾¤ä½“ç”Ÿç‰©å­¦è¿‡ç¨‹ã€‚ä¹Ÿå°±æ˜¯æˆ‘ä»¬ç»å¸¸è¯´çš„æ‹Ÿæ—¶åºï¼ˆpseudotimeï¼‰åˆ†æï¼Œåˆç§°ç»†èƒè½¨è¿¹ï¼ˆcell trajectoryï¼‰åˆ†æã€‚é€šè¿‡æ‹Ÿæ—¶åºåˆ†æå¯ä»¥æ¨æ–­å‡ºå‘è‚²è¿‡ç¨‹ç»†èƒçš„åˆ†åŒ–è½¨è¿¹æˆ–ç»†èƒäºšå‹çš„æ¼”åŒ–è¿‡ç¨‹ï¼Œä¸»è¦åŸºäºå…³é”®åŸºå› çš„è¡¨è¾¾æ¨¡å¼ï¼Œé€šè¿‡å­¦ä¹ æ¯ä¸ªç»†èƒå¿…é¡»ç»å†çš„åŸºå› è¡¨è¾¾å˜åŒ–çš„åºåˆ—ï¼Œæ ¹æ®æ‹Ÿæ—¶é—´å€¼ä¸­å¯¹å•ä¸ªç»†èƒè¿›è¡Œæ’åºï¼Œæ¨¡æ‹Ÿå‡ºæ—¶é—´å‘è‚²è¿‡ç¨‹çš„åŠ¨æ€å˜åŒ–ã€‚
<br> 
<b><mark>ä¼ªæ—¶é—´æ˜¯è¡¡é‡å•ä¸ªç»†èƒé€šè¿‡ç»†èƒåˆ†åŒ–ç­‰è¿‡ç¨‹å–å¾—å¤šå°‘è¿›å±•çš„æŒ‡æ ‡ </mark></b>Monocle ä¸æ˜¯è·Ÿè¸ªä½œä¸ºæ—¶é—´å‡½æ•°çš„è¡¨è¾¾å˜åŒ–ï¼Œè€Œæ˜¯è·Ÿè¸ªä½œä¸ºæ²¿è½¨è¿¹è¿›å±•çš„å‡½æ•°çš„å˜åŒ–ï¼Œæˆ‘ä»¬ç§°ä¹‹ä¸ºä¼ªæ—¶é—´ã€‚ä¼ªæ—¶é—´æ˜¯ä¸€ä¸ªæŠ½è±¡çš„è¿›ç¨‹å•ä½ï¼šå®ƒåªæ˜¯ä¸€ä¸ªå•å…ƒæ ¼å’Œè½¨è¿¹èµ·ç‚¹ä¹‹é—´çš„è·ç¦»ï¼Œæ²¿ç€æœ€çŸ­è·¯å¾„æµ‹é‡ã€‚è½¨è¿¹çš„æ€»é•¿åº¦æ˜¯æ ¹æ®ç»†èƒä»èµ·å§‹çŠ¶æ€ç§»åŠ¨åˆ°ç»“æŸçŠ¶æ€æ—¶æ‰€ç»å†çš„è½¬å½•å˜åŒ–æ€»é‡æ¥å®šä¹‰çš„ã€‚

>æ­¥éª¤

1. å¯¼å…¥æ•°æ®ï¼Œåˆ›å»ºCellDataSet ç±»çš„å¯¹è±¡
* å¯¼å…¥æ•°æ®ï¼Œåˆ›å»ºCellDataSet ç±»çš„å¯¹è±¡
```r
HSMM_expr_matrix <- read.table("fpkm_matrix.txt")
HSMM_sample_sheet <- read.delim("cell_sample_sheet.txt")
HSMM_gene_annotation <- read.delim("gene_annotations.txt")

pd <- new("AnnotatedDataFrame", data = HSMM_sample_sheet)
fd <- new("AnnotatedDataFrame", data = HSMM_gene_annotation)
HSMM <- newCellDataSet(as.matrix(HSMM_expr_matrix),
    phenoData = pd, featureData = fd)

rm(list=ls())
# åŠ è½½éœ€è¦çš„RåŒ…
library(Seurat)
library(monocle)
# è®¾ç½®cell rangerè¾“å‡ºç»“æœç›®å½•
input_dir <- "/Users/hecate/ç ”ä¸€/blood_vessels_scRNA/matrix/BK-1111_matrix_10X"
# è¯»å–æ•°æ®
pbmc <- Read10X(input_dir)
pbmc <- CreateSeuratObject(pbmc, project = "pbmc10k", min.cells = 3, min.features = 200)

data <- as(as.matrix(pbmc@assays$RNA@data), 'sparseMatrix')

pd <- new('AnnotatedDataFrame', data = pbmc@meta.data)

fData <- data.frame(gene_short_name = row.names(data), row.names = row.names(data))
fd <- new('AnnotatedDataFrame', data = fData)

#Construct monocle cds
cds <- newCellDataSet(data,
                      phenoData = pd,
                      featureData = fd,
                      lowerDetectionLimit = 0.5,
                      expressionFamily = negbinomial.size())
```
<b>å¯¼å…¥å’Œå¯¼å‡ºSeuratæˆ–è€…scateråŒ…çš„æ•°æ®<font color=#FF0000>ï¼ˆåœ¨åšæ‹Ÿæ—¶åºä¹‹å‰ï¼Œæœ€å¥½å…ˆè·‘å®Œseuratæ ‡å‡†æµç¨‹ï¼Œå¹¶æ³¨é‡Šå¥½ç»†èƒç±»å‹ã€‚ï¼‰</b></font>
```r
HSMM_expr_matrix <- read.table("fpkm_matrix.txt")
HSMM_sample_sheet <- read.delim("cell_sample_sheet.txt")
HSMM_gene_annotation <- read.delim("gene_annotations.txt")

pd <- new("AnnotatedDataFrame", data = HSMM_sample_sheet)
fd <- new("AnnotatedDataFrame", data = HSMM_gene_annotation)
HSMM <- newCellDataSet(as.matrix(HSMM_expr_matrix),
    phenoData = pd, featureData = fd)
# ä¸Seuratï¼Œscaterç­‰RåŒ…çš„äº¤äº’
## å¯¼å…¥ importCDS()
# Where 'data_to_be_imported' can either be a Seurat object
# or an SCESet.

importCDS(data_to_be_imported)

# We can set the parameter 'import_all' to TRUE if we'd like to
# import all the slots from our Seurat object or SCESet.
# (Default is FALSE or only keep minimal dataset)

importCDS(data_to_be_imported, import_all = TRUE)

## å¯¼å‡º exportCDS()
# To convert to Seurat object
lung_seurat <- exportCDS(lung, 'Seurat')

# To convert to SCESet
lung_SCESet <- exportCDS(lung, 'Scater')
```

* é€‰æ‹©æ•°æ®çš„åˆ†å¸ƒç±»å‹â—â—â—ï¸
Monocle é€‚ç”¨äºç›¸å¯¹è¡¨è¾¾æ•°æ®å’ŒåŸºäºè®¡æ•°çš„æµ‹é‡ï¼ˆä¾‹å¦‚ UMIï¼‰ã€‚ä¸€èˆ¬æ¥è¯´ï¼Œå®ƒæœ€é€‚ç”¨äºè½¬å½•æœ¬è®¡æ•°æ•°æ®ï¼Œå°¤å…¶æ˜¯ UMI æ•°æ®ã€‚<font color=#FF0000>FPKM/TPM å€¼é€šå¸¸å‘ˆå¯¹æ•°æ­£æ€åˆ†å¸ƒï¼Œè€Œ UMI æˆ–è¯»å–è®¡æ•°æ›´é€‚åˆä½¿ç”¨è´ŸäºŒé¡¹å¼åˆ†å¸ƒå»ºæ¨¡ã€‚</font>
```r
HSMM <- newCellDataSet(count_matrix,
                phenoData = pd,
                featureData = fd,
                expressionFamily=negbinomial.size())
```
<table>
<tr><td>negbinomial.size()</td><td>UMIã€æ¥è‡ªæºå…¥å®éªŒçš„è½¬å½•æœ¬è®¡æ•°æˆ–relative2abs()åŸå§‹è¯»å–è®¡æ•°</td></tr>
<tr><td>negbinomial()</td><td>UMIã€æ¥è‡ªæºå…¥å®éªŒçš„è½¬å½•æœ¬è®¡æ•°æˆ–relative2absåŸå§‹è¯»å–è®¡æ•°</td></tr>
<tr><td>tobit()</td><td>FPKMã€TPM</td></tr>
<tr><td>gaussianff()</td><td>å¯¹æ•°è½¬æ¢çš„ FPKM/TPMï¼Œæ¥è‡ªå•ç»†èƒ qPCR çš„ Ct å€¼</td></tr>
</table>

* ä½¿ç”¨å¤§å‹æ•°æ®é›†ï¼ˆæ¨èï¼‰
ä½¿ç”¨*ç¨€ç–çŸ©é˜µ*å¯ä»¥å¸®åŠ©æ‚¨åœ¨å…¸å‹è®¡ç®—æœºä¸Šå¤„ç†å¤§é‡æ•°æ®é›†ã€‚æˆ‘ä»¬é€šå¸¸å»ºè®®å¤§å¤šæ•°ç”¨æˆ·ä½¿ç”¨ sparseMatricesï¼Œå› ä¸ºå®ƒå¯ä»¥åŠ é€Ÿè®¸å¤šè®¡ç®—ï¼Œå³ä½¿å¯¹äºè¾ƒå°è§„æ¨¡çš„æ•°æ®é›†ä¹Ÿæ˜¯å¦‚æ­¤ã€‚
```r
HSMM <- newCellDataSet(as(umi_matrix, "sparseMatrix"),
                phenoData = pd,
                featureData = fd,
                lowerDetectionLimit = 0.5,
                expressionFamily = negbinomial.size())
```
â—â—â—ï¸è®¸å¤š RNA-Seq æµç¨‹ï¼ˆåŒ…æ‹¬ CellRangerï¼‰çš„è¾“å‡ºå·²ç»æ˜¯ç¨€ç–çŸ©é˜µæ ¼å¼ï¼ˆä¾‹å¦‚ MTXï¼‰ã€‚è¿™æ—¶ç›´æ¥å°†å…¶ä¼ é€’ç»™newCellDataSetï¼Œè€Œæ— éœ€å…ˆå°†å…¶è½¬æ¢ä¸ºå¯†é›†çŸ©é˜µï¼ˆé€šè¿‡as.matrix()ï¼Œå› ä¸ºè¿™å¯èƒ½ä¼šè¶…å‡ºå¯ç”¨å†…å­˜ã€‚
<br>å¦‚æœä½¿ç”¨çš„æ˜¯<font color=#0000FF>10X Genomics æ•°æ®å¹¶ä¸”æ­£åœ¨ä½¿ç”¨cellrangerRkitï¼Œä½¿ç”¨ä¸‹åˆ—ä»£ç è¿›è¡ŒåŠ è½½è¯¥æ•°æ®ã€‚</font>
```r
cellranger_pipestance_path <- "/path/to/your/pipeline/output/directory"
gbm <- load_cellranger_matrix(cellranger_pipestance_path)

fd <- fData(gbm)

# The number 2 is picked arbitrarily in the line below.
# Where "2" is placed you should place the column number that corresponds to your
# featureData's gene short names.

colnames(fd)[2] <- "gene_short_name"

gbm_cds <- newCellDataSet(exprs(gbm),
                  phenoData = new("AnnotatedDataFrame", data = pData(gbm)),
                  featureData = new("AnnotatedDataFrame", data = fd),
                  lowerDetectionLimit = 0.5,
                  expressionFamily = negbinomial.size())
```
*  ğŸ”ºä¼°è®¡å°ºå¯¸å› å­å’Œåˆ†æ•£<font color=#FF0000>ï¼ˆéœ€è¦ï¼‰</font>
<br>å°ºå¯¸å› å­å¸®åŠ©æˆ‘ä»¬å¯¹è·¨ç»†èƒæ¢å¤çš„ mRNA å·®å¼‚è¿›è¡Œæ ‡å‡†åŒ–ã€‚
<br>â€œåˆ†æ•£â€å€¼å°†å¸®åŠ©æˆ‘ä»¬ç¨åè¿›è¡Œå·®å¼‚è¡¨è¾¾åˆ†æã€‚
<br><b>è¦æ±‚ï¼š</b>
`stimateSizeFactors()`å’Œ`estimateDispersions()`ä»…åœ¨ä½¿ç”¨å¸¦æœ‰`negbinomial()``ornegbinomial.size()`çš„ `CellDataSet` æ—¶æ‰æœ‰æ•ˆã€‚
```r
HSMM <- estimateSizeFactors(HSMM)
HSMM <- estimateDispersions(HSMM)
```
* è¿‡æ»¤ä½è´¨é‡ç»†èƒï¼ˆæ¨èï¼‰
```r
# æŸ¥çœ‹ç»†èƒä¸­ mRNA æ€»æ•°çš„åˆ†å¸ƒ
pData(HSMM)$Total_mRNAs <- Matrix::colSums(exprs(HSMM))

HSMM <- HSMM[,pData(HSMM)$Total_mRNAs < 1e6]

upper_bound <- 10^(mean(log10(pData(HSMM)$Total_mRNAs)) +
            2*sd(log10(pData(HSMM)$Total_mRNAs)))
lower_bound <- 10^(mean(log10(pData(HSMM)$Total_mRNAs)) -
            2*sd(log10(pData(HSMM)$Total_mRNAs)))

qplot(Total_mRNAs, data = pData(HSMM), color = Hours, geom =
"density") +
geom_vline(xintercept = lower_bound) +
geom_vline(xintercept = upper_bound)
```
![](images/2021-10-25-15-16-27.png)
```r
HSMM <- HSMM[,pData(HSMM)$Total_mRNAs > lower_bound &
      pData(HSMM)$Total_mRNAs < upper_bound]
HSMM <- detectGenes(HSMM, min_expr = 0.1) #è®¡ç®—æ¯ä¸ªåŸºå› åœ¨å¤šå°‘ç»†èƒä¸­è¡¨è¾¾

# éªŒè¯CellDataSet æ˜¯å¦éµå¾ªå¤§è‡´å¯¹æ•°æ­£æ€åˆ†å¸ƒ
# Log-transform each value in the expression matrix.
L <- log(exprs(HSMM[expressed_genes,]))

# Standardize each gene, so that they are all on the same scale,
# Then melt the data with plyr so we can plot it easily
melted_dens_df <- melt(Matrix::t(scale(Matrix::t(L))))

# Plot the distribution of the standardized gene expression values.
qplot(value, geom = "density", data = melted_dens_df) +
stat_function(fun = dnorm, size = 0.5, color = 'red') +
xlab("Standardized log(FPKM)") +
ylab("Density")
```
![](images/2021-10-25-15-16-45.png)

2. èšç±»å’Œç»†èƒç±»å‹æ³¨é‡Šï¼ˆä¸€èˆ¬æ¨èä½¿ç”¨Seuratçš„æ ‡å‡†æµç¨‹ï¼‰
* <b>Step1: é€‰æ‹©å®šä¹‰ç»†èƒè¿›ç¨‹çš„åŸºå› ï¼ˆç‰¹å¾é€‰æ‹©ï¼‰</b>
```r
# é€‰æ‹©çš„åŸºå› éšç€æˆ‘ä»¬æ­£åœ¨ç ”ç©¶çš„è¿‡ç¨‹çš„è¿›å±•è€Œå¢åŠ ï¼ˆæˆ–å‡å°‘ï¼‰è¡¨è¾¾ï¼ˆå·®å¼‚è¡¨è¾¾çš„åŸºå› ï¼‰
## åˆ†ç¦»ä¸€ç»„æ’åºåŸºå› çš„ä¸€ç§æœ‰æ•ˆæ–¹æ³•æ˜¯ç®€å•åœ°å°†è¿‡ç¨‹å¼€å§‹æ—¶æ”¶é›†çš„ç»†èƒä¸è¿‡ç¨‹ç»“æŸæ—¶æ”¶é›†çš„ç»†èƒè¿›è¡Œæ¯”è¾ƒï¼Œå¹¶æ‰¾åˆ°å·®å¼‚è¡¨è¾¾çš„åŸºå› 
diff_test_res <- differentialGeneTest(HSMM_myo[expressed_genes,],
              fullModelFormulaStr = "~Media")
ordering_genes <- row.names (subset(diff_test_res, qval < 0.01))
# æ— æ—¶é—´åºåˆ—æ•°æ®æ—¶
HSMM_myo <- setOrderingFilter(HSMM_myo, ordering_genes)
plot_ordering_genes(HSMM_myo)
```
* <b>Step2:é™ä½æ•°æ®ç»´åº¦</b> 
<br>DDRTree: å°†é«˜ç»´ç©ºé—´çš„æ•°æ®ç‚¹é™è‡³äºŒç»´ï¼Œæ–¹ä¾¿å¯è§†åŒ–
```r
HSMM_myo <- reduceDimension(HSMM_myo, max_components = 2,
    method = 'DDRTree')
```
â—â—â—ï¸Monocle3ä½¿ç”¨çš„é™ç»´æ–¹æ³•ä¸ºUMAPå’ŒtSNE

* <b>Step3: æ²¿è½¨è¿¹å¯¹ç»†èƒè¿›è¡Œæ’åº</b>
<br>è½¨è¿¹å…·æœ‰æ ‘çŠ¶ç»“æ„ã€‚æˆ‘ä»¬å¯ä»¥çœ‹åˆ°ï¼Œåœ¨æ—¶é—´é›¶æ”¶é›†çš„ç»†èƒä½äºæ ‘å°–ä¹‹ä¸€é™„è¿‘ï¼Œè€Œå…¶ä»–ç»†èƒåˆ†å¸ƒåœ¨ä¸¤ä¸ªâ€œæ ‘æâ€ä¹‹é—´ã€‚Monocle ä¸çŸ¥é“å…ˆéªŒåœ° å°†æ ‘çš„å“ªä¸ªè½¨è¿¹ç§°ä¸ºâ€œå¼€å§‹â€ï¼Œå› æ­¤å¯¹`orderCells`ä½¿ç”¨`root_state`å‚æ•°å†æ¬¡è°ƒç”¨ä»¥æŒ‡å®šå¼€å§‹ã€‚

```r
HSMM_myo <- orderCells(HSMM_myo)
## â—ï¸è®¾å®šæ ¹ç»“ç‚¹ï¼Œéœ€è¦æ ¹æ®åç»­çš„å›¾æ¥è¿›ä¸€æ­¥ç¡®å®šæœ€ä½³çš„æ ¹ç»“ç‚¹
mycds <- orderCells(mycds, root_state = 2)
## ç»“æœå¯è§†åŒ–
## Hoursè½¨è¿¹åˆ†å¸ƒå›¾
plot_cell_trajectory(HSMM_myo, color_by = "Hours")
```
![](images/2021-10-25-15-19-54.png)
```r
# åœ¨æ—¶é—´é›¶æ”¶é›†çš„ç»†èƒä½äºæ ‘å°–ä¹‹ä¸€é™„è¿‘ï¼Œè€Œå…¶ä»–ç»†èƒåˆ†å¸ƒåœ¨ä¸¤ä¸ªâ€œæ ‘æâ€ä¹‹é—´ã€‚Monocle ä¸çŸ¥é“å…ˆéªŒåœ°å°†æ ‘çš„å“ªä¸ªè½¨è¿¹ç§°ä¸ºâ€œå¼€å§‹â€ï¼Œå› æ­¤æˆ‘ä»¬éœ€è¦å†æ¬¡è°ƒå‚ä»¥æŒ‡å®šå¼€å§‹ã€‚
# é€šè¿‡â€œstateâ€ä¸ºç»†èƒç€è‰²
plot_cell_trajectory(HSMM_myo, color_by = "State")
```
![](images/2021-10-25-15-20-16.png)
```r
# é€šè¿‡è¯†åˆ«åŒ…å«ä»æ—¶é—´é›¶å¼€å§‹çš„å¤§éƒ¨åˆ†ç»†èƒçš„çŠ¶æ€
GM_state <- function(cds){
  if (length(unique(pData(cds)$State)) > 1){
    T0_counts <- table(pData(cds)$State, pData(cds)$Hours)[,"0"]
    return(as.numeric(names(T0_counts)[which
          (T0_counts == max(T0_counts))]))
  } else {
    return (1)
  }
}
##Pseudotimeè½¨è¿¹å›¾
HSMM_myo <- orderCells(HSMM_myo, root_state = GM_state(HSMM_myo))
plot_cell_trajectory(HSMM_myo, color_by = "Pseudotime")
```
![](images/2021-10-25-15-24-53.png)
```r
# â€œåˆ†é¢â€è½¨è¿¹å›¾
plot_cell_trajectory(HSMM_myo, color_by = "State") +
    facet_wrap(~State, nrow = 1)
```
![](images/2021-10-25-15-25-09.png)
```r
# æ— æ—¶é—´åºåˆ—ï¼Œéœ€å…ˆéªŒçŸ¥è¯†ï¼Œæ ¹æ®æŸäº›æ ‡è®°åŸºå› çš„è¡¨è¾¾ä½ç½®æ¥è®¾ç½®æ ¹
blast_genes <- row.names(subset(fData(HSMM_myo),
gene_short_name %in% c("CCNB2", "MYOD1", "MYOG")))
plot_genes_jitter(HSMM_myo[blast_genes,],
    grouping = "State",
min_expr = 0.1)
```
![](images/2021-10-25-15-25-26.png)
```r
# ä¸ºäº†ç¡®è®¤æ’åºæ˜¯æ­£ç¡®çš„ï¼Œç»˜åˆ¶é€‰æ‹©çš„åŸºå› çš„è¡¨è¾¾æ°´å¹³
HSMM_expressed_genes <-  row.names(subset(fData(HSMM_myo),
num_cells_expressed >= 10))
HSMM_filtered <- HSMM_myo[HSMM_expressed_genes,]
my_genes <- row.names(subset(fData(HSMM_filtered),
          gene_short_name %in% c("CDK1", "MEF2C", "MYH3")))
cds_subset <- HSMM_filtered[my_genes,]
plot_genes_in_pseudotime(cds_subset, color_by = "Hours")
```
![](images/2021-10-25-15-25-47.png)

<b>è¡¥å……ï¼š</b>å¯ä»¥é€šè¿‡é€‰æ‹©å‘è‚²å·®å¼‚è¡¨è¾¾åŸºå› ï¼Œé€‰æ‹©clusterså·®å¼‚è¡¨è¾¾åŸºå› ï¼Œé€‰æ‹©ç¦»æ•£ç¨‹åº¦é«˜çš„åŸºå› ã€‚è‡ªå®šä¹‰å‘è‚²markeråŸºå› æ¥é€‰æ‹©åŸºå› ï¼Œå‰ä¸‰ä¸ªä¸ºæ— ç›‘ç£æ–¹æ³•ï¼Œæœ€åä¸€ä¸ªä¸ºåŠç›‘ç£åˆ†æï¼Œéœ€è¦å…ˆéªŒçŸ¥è¯†ã€‚
```r
##ä½¿ç”¨seuraté€‰æ‹©çš„é«˜å˜åŸºå› âš ï¸
express_genes <- VariableFeatures(pbmc)
cds <- setOrderingFilter(cds, express_genes)
plot_ordering_genes(cds)
##ä½¿ç”¨clusterså·®å¼‚è¡¨è¾¾markeråŸºå› 
deg.cluster <- FindAllMarkers(pbmc)
express_genes <- subset(deg.cluster,p_val_adj<0.05)$gene
cds <- setOrderingFilter(cds, express_genes)
plot_ordering_genes(cds)
##ä½¿ç”¨monocleé€‰æ‹©çš„é«˜å˜åŸºå› âš ï¸
disp_table <- dispersionTable(cds)
disp.genes <- subset(disp_table, mean_expression >= 0.1 & dispersion_empirical >= 1 * dispersion_fit)$gene_id
cds <- setOrderingFilter(cds, disp.genes)
plot_ordering_genes(cds)
```
> Monocle3
å…¶ä¸­ï¼ŒMonocle3ä¸Monocle2çš„åŒºåˆ«åœ¨äºï¼š
* ä»UMAPå›¾è¯†åˆ«å‘è‚²è½¨è¿¹ï¼Œå¯ä»¥ç»§æ‰¿Seuratçš„è´¨æ§ã€æ‰¹æ¬¡æ ¡æ­£å’Œé™ç»´åˆ†æç»“æœï¼Œå®ç°â€œä¸€å¼ å›¾â€å±•ç°ç»†èƒçš„èšç±»ã€é‰´å®šå’Œè½¨è¿¹åˆ†æç»“æœã€‚
* è‡ªåŠ¨å¯¹UMAPå›¾åˆ†åŒºï¼ˆpartitionï¼‰ï¼Œå¯ä»¥é€‰æ‹©å¤šä¸ªèµ·ç‚¹ï¼Œè½¨è¿¹åˆ†æç®—æ³•çš„é€»è¾‘æ›´ç¬¦åˆç”Ÿç‰©å­¦ç°å®ã€‚
* å°†æ‹Ÿåˆä¸€ä¸ªä¸»å›¾åœ¨æ¯ä¸ªåˆ†åŒºå†…ä½¿ç”¨learn_graph()å‡½æ•°ï¼Œä¸ºäº†å°†ç»†èƒæ’åˆ—æ•´é½ï¼Œæˆ‘ä»¬é€šè¿‡é€‰æ‹©æˆ‘ä»¬æ ‡è®°ä¸ºè½¨è¿¹â€œæ ¹â€çš„å›¾å½¢åŒºåŸŸæ¥ç¡®å®šç”Ÿç‰©è¿‡ç¨‹çš„â€œå¼€å§‹â€åœ¨å“ªé‡Œã€‚

```r
library(monocle3)
library(Seurat)
library(ggplot2)
library(ggsci)

#
# 1. create
#
# pbmc.data <- Read10X(data.dir = "~/project/10x_sc/data/pbmc3k/filtered_gene_bc_matrices_h19")
pbmc.data <- Read10X(data.dir = "~/filtered_gene_bc_matrices_h19")
satija <- readRDS(file = "/home/renjun/filtered_gene_bc_matrices_h19/pbmc3k_final.rds")

expr_matrix <- as.matrix(pbmc.data)
sample_sheet <- data.frame(row.names = colnames(pbmc.data),
                           id = colnames(pbmc.data),
                           stringsAsFactors = FALSE)
gene_annotation <- data.frame(row.names = rownames(pbmc.data),
                              id = rownames(pbmc.data),
                              gene_short_name = rownames(pbmc.data),
                              stringsAsFactors = FALSE)
cds <- new_cell_data_set(expr_matrix,
                         cell_metadata = sample_sheet,
                         gene_metadata = gene_annotation)
# cds <- load_mm_data(mat_path = "~/filtered_gene_bc_matrices_h19/matrix.mtx", 
#                     feature_anno_path = "~/filtered_gene_bc_matrices_h19/genes.tsv", 
#                     cell_anno_path = "~/filtered_gene_bc_matrices_h19/barcodes.tsv")

#
# 2. preprocessing
#
cds <- preprocess_cds(cds, num_dim = 50, norm_method = "log", scaling = TRUE)
ggsave("pv_variance.png", plot_pc_variance_explained(cds), width = 5, height = 5)

#
# 3. reduce dimension
#
cds <- reduce_dimension(cds, max_components = 2, reduction_method = "UMAP", preprocess_method = "PCA")
cds <- reduce_dimension(cds, max_components = 2, reduction_method = "tSNE", preprocess_method = "PCA")
# cds <- align_cds(cds, alignment_group = "batch", residual_model_formula_str = "~ bg.300.loading + bg.400.loading")
# marker <- c("MS4A1", "GNLY", "CD3E", "CD14", "FCER1A", "FCGR3A", 
#            "LYZ", "PPBP", "CD8A","IL7R","CCR7","S100A4")

#
# 4. clustering
#
cds <- cluster_cells(cds, reduction_method = "UMAP", k = 20, cluster_method = "leiden", resolution = 1e-4) #default
cds <- cluster_cells(cds, reduction_method = "tSNE", k = 20, cluster_method = "leiden", resolution = 1e-2)
cds <- cluster_cells(cds, reduction_method = "PCA", k = 20, cluster_method = "louvain")
# cds@clusters@listData$PCA$clusters

# plot cell with cluster
p1 <- plot_cells(cds, reduction_method = "UMAP", color_cells_by = "cluster", 
                 cell_size = 2.3, alpha = 0.5, group_label_size = 5) + 
  scale_colour_d3("category20") +  rj.ftheme
p2 <- plot_cells(cds, reduction_method = "tSNE", color_cells_by = "cluster", 
                 cell_size = 2.3, alpha = 0.5, group_label_size = 5) + 
  scale_colour_d3("category20") +  rj.ftheme
p3 <- plot_cells(cds, reduction_method = "PCA", color_cells_by = "cluster", 
                 cell_size = 2.3, alpha = 0.5, group_label_size = 5) + 
  scale_colour_d3("category20") +  rj.ftheme

ggsave("2d_cluster.png", CombinePlots(list(p1,p2,p3), ncol = 3), width = 12, height = 4)

# plot cell with partition
p1 <- plot_cells(cds, reduction_method = "UMAP", color_cells_by = "partition", 
                 cell_size = 2.3, alpha = 0.5, group_label_size = 5) + 
  scale_colour_d3("category20") +  rj.ftheme
p2 <- plot_cells(cds, reduction_method = "tSNE", color_cells_by = "partition", 
                 cell_size = 2.3, alpha = 0.5, group_label_size = 5) + 
  scale_colour_d3("category20") +  rj.ftheme
p3 <- plot_cells(cds, reduction_method = "PCA", color_cells_by = "partition", 
                 cell_size = 2.3, alpha = 0.5, group_label_size = 5) + 
  scale_colour_d3("category20") +  rj.ftheme

ggsave("2d_partition.png", CombinePlots(list(p1,p2,p3), ncol = 3), width = 12, height = 4)

#
# 5. find marker
#
marker_test_res <- top_markers(cds, group_cells_by = "cluster", reduction_method = "UMAP", reference_cells = 1000, cores = 1)

top_specific_markers <- marker_test_res %>%
  filter(fraction_expressing >= 0.10) %>%
  group_by(cell_group) %>%
  top_n(10, pseudo_R2)

top_specific_marker_ids <- unique(top_specific_markers %>% pull(gene_id))

p <- plot_genes_by_group(cds,
                         top_specific_marker_ids,
                         group_cells_by = "cluster",
                         reduction_method = "UMAP",
                         ordering_type = "maximal_on_diag",
                         max.size = 3)
ggsave("top_gene10_heatmap.png", p, width = 6, height = 8)

#
# 6. choose cells for a subset
#
cds_subset <- choose_cells(cds)

#
# 7. learn the trajectory
#
cds <- learn_graph(cds)

# Passing the programatically selected root node to order_cells() via the root_pr_nodeargument yields:
cds <- order_cells(cds)
p <- plot_cells(cds,
                color_cells_by = "pseudotime",
                reduction_method = "UMAP",
                label_groups_by_cluster=FALSE,
                label_leaves=TRUE,
                label_branch_points=FALSE) + rj.ftheme
ggsave("pseudotime2.png", p, width = 6, height = 6)

save(cds, file = "cds.rda")
```
### åŸºå› æ°´å¹³
#### åŸºå› çš„å·®å¼‚è¡¨è¾¾åˆ†æ
[â—ï¸â—ï¸â—ï¸è¯¥æ­¥éª¤ä¸€èˆ¬åœ¨èšç±»åˆ†æä¹‹åï¼Œç»†èƒç±»å‹æ³¨é‡Šä¹‹å‰è¿›è¡Œ]

##### å¯»æ‰¾å·®å¼‚è¡¨è¾¾çš„ç‰¹å¾åŸºå› ï¼ˆç°‡ç”Ÿç‰©æ ‡å¿—ç‰©ï¼‰
> SeuratåŒ…

1. è¿è¡Œéçº¿æ€§é™ç»´â—ï¸â—ï¸â—ï¸

Seuratæä¾›å‡ ä¸ªéçº¿æ€§é™ç»´æ–¹æ³•ï¼Œå¦‚tSNEå’ŒUMAPï¼Œæ¥å¯è§†åŒ–å’Œæ¢ç´¢è¿™äº›æ•°æ®é›†ã€‚è¿™äº›ç®—æ³•çš„ç›®æ ‡æ˜¯å­¦ä¹ æ•°æ®çš„åº•å±‚æµå½¢ï¼Œä»¥ä¾¿å°†ç›¸ä¼¼çš„ç»†èƒæ”¾åœ¨ä½ç»´ç©ºé—´ä¸­ã€‚ä¸Šé¢ç¡®å®šçš„åŸºäºå›¾å½¢çš„ç°‡ä¸­çš„ç»†èƒåº”è¯¥å…±åŒå®šä½åœ¨è¿™äº›é™ç»´å›¾ä¸­ã€‚ä½œä¸º UMAP å’Œ tSNE çš„è¾“å…¥ï¼Œæˆ‘ä»¬å»ºè®®ä½¿ç”¨ç›¸åŒçš„ PC ä½œä¸ºèšç±»åˆ†æçš„è¾“å…¥ã€‚ç”±äºt-SNEæ‰€ä½¿ç”¨çš„æ—¶é—´è¾ƒé•¿ï¼Œä¸€èˆ¬ä½¿ç”¨UMAPã€‚
```r
# If you haven't installed UMAP, you can do so via reticulate::py_install(packages =
# 'umap-learn')
pbmc <- RunUMAP(pbmc, dims = 1:10)
# note that you can set `label = TRUE` or use the LabelClusters function to help label
# individual clusters
DimPlot(pbmc, reduction = "umap")
# save the data
saveRDS(pbmc, file = "../output/pbmc_tutorial.rds")
```
![](images/2021-10-25-15-40-12.png)
2. markeråŸºå› çš„è¯†åˆ«
<br>æ‰¾åˆ°é€šè¿‡å·®å¼‚è¡¨è¾¾å®šä¹‰èšç±»çš„æ ‡è®°åŸºå› ã€‚é»˜è®¤æƒ…å†µä¸‹ï¼Œ`ident.1`ä¸æ‰€æœ‰å…¶ä»–ç»†èƒç›¸æ¯”ï¼Œå®ƒè¯†åˆ«å•ä¸ªç°‡ï¼ˆåœ¨`ident.1`ä¸­æŒ‡å®šï¼‰çš„é˜³æ€§å’Œé˜´æ€§æ ‡è®°ã€‚`FindAllMarkers()`ä¸ºæ‰€æœ‰ç°‡è‡ªåŠ¨æ‰§è¡Œæ­¤è¿‡ç¨‹ï¼Œä½†ä¹Ÿå¯ä»¥æµ‹è¯•ç°‡ä¸ç°‡ä¹‹é—´çš„å¯¹æ¯”ï¼Œæˆ–é’ˆå¯¹æ‰€æœ‰ç»†èƒè¿›è¡Œæµ‹è¯•ã€‚
<br>è¯¥`min.pct`å‚æ•°è¦æ±‚åœ¨ä¸¤ç»„ç»†èƒä¸­çš„ä»»ä½•ä¸€ç»„ä¸­ä»¥æœ€å°ç™¾åˆ†æ¯”æ£€æµ‹åˆ°ä¸€ä¸ªåŸºå› ï¼Œè€Œ `thresh.test`å‚æ•°è¦æ±‚ä¸€ä¸ªåŸºå› åœ¨ä¸¤ç»„ä¹‹é—´å·®å¼‚è¡¨è¾¾ï¼ˆå¹³å‡ï¼‰ä¸€å®šé‡ã€‚å¯ä»¥å°†è¿™ä¸¤ä¸ªéƒ½è®¾ç½®ä¸º 0ï¼Œä½†æ—¶é—´ä¼šæ˜¾ç€å¢åŠ ï¼Œå› ä¸ºè¿™å°†æµ‹è¯•å¤§é‡ä¸å¤ªå¯èƒ½å…·æœ‰é«˜åº¦æ­§è§†æ€§çš„åŸºå› ã€‚ä½œä¸ºåŠ é€Ÿè¿™äº›è®¡ç®—çš„å¦ä¸€ä¸ªé€‰é¡¹ï¼Œ`max.cells.per.ident`å¯ä»¥è®¾ç½®ã€‚
<br>Seurat æœ‰å‡ ç§å·®å¼‚è¡¨è¾¾æµ‹è¯•ï¼Œå¯ä»¥ä½¿ç”¨ test.use å‚æ•°è¿›è¡Œè®¾ç½®ï¼Œæ¥ç€å¯è§†åŒ–ã€‚
<br>ç›®å‰æ”¯æŒä»¥ä¸‹å·®å¼‚è¡¨è¾¾æµ‹è¯•ï¼š
<table>
   <tr>
      <td>â€œwilcoxâ€ï¼šWilcoxon ç§©å’Œæ£€éªŒï¼ˆé»˜è®¤ï¼‰</td>
   </tr>
   <tr>
      <td>â€œbimodâ€ï¼šå•ç»†èƒç‰¹å¾è¡¨è¾¾çš„ä¼¼ç„¶æ¯”æ£€éªŒã€é»˜è®¤ã€‘</td>
   </tr>
   <tr>
      <td>â€œrocâ€ï¼šæ ‡å‡† AUC åˆ†ç±»å™¨</td>
   </tr>
   <tr>
      <td>â€œtâ€ï¼šå­¦ç”Ÿçš„ t æ£€éªŒ</td>
   </tr>
   <tr>
      <td>â€œæ³Šæ¾â€ï¼šå‡è®¾æ½œåœ¨è´ŸäºŒé¡¹å¼åˆ†å¸ƒçš„ä¼¼ç„¶æ¯”æ£€éªŒã€‚ä»…ç”¨äºåŸºäº UMI çš„æ•°æ®é›†</td>
   </tr>
   <tr>
      <td>â€œnegbinomâ€ï¼šå‡è®¾æ½œåœ¨è´ŸäºŒé¡¹å¼åˆ†å¸ƒçš„ä¼¼ç„¶æ¯”æ£€éªŒã€‚ä»…ç”¨äºåŸºäº UMI çš„æ•°æ®é›†</td>
   </tr>
   <tr>
      <td>â€œLRâ€ï¼šä½¿ç”¨é€»è¾‘å›å½’æ¡†æ¶æ¥ç¡®å®šå·®å¼‚è¡¨è¾¾çš„åŸºå› ã€‚æ„å»ºé€»è¾‘å›å½’æ¨¡å‹ï¼Œåˆ†åˆ«åŸºäºæ¯ä¸ªç‰¹å¾é¢„æµ‹ç»„æˆå‘˜èº«ä»½ï¼Œå¹¶å°†å…¶ä¸å…·æœ‰ä¼¼ç„¶æ¯”æ£€éªŒçš„ç©ºæ¨¡å‹è¿›è¡Œæ¯”è¾ƒã€‚</td>
   </tr>
   <tr>
      <td>â€œMASTâ€ï¼šå°†ç»†èƒæ£€æµ‹ç‡è§†ä¸ºåå˜é‡çš„ GLM æ¡†æ¶</td>
   </tr>
   <tr>
      <td>â€œDESeq2â€ï¼šDE åŸºäºä½¿ç”¨è´ŸäºŒé¡¹åˆ†å¸ƒçš„æ¨¡å‹</td>
   </tr>
   <tr>
      <td></td>
   </tr>
</table>

```r
# find all markers of cluster 2
cluster2.markers <- FindMarkers(pbmc, ident.1 = 2, min.pct = 0.25)
head(cluster2.markers, n = 5)
# find all markers distinguishing cluster 5 from clusters 0 and 3
cluster5.markers <- FindMarkers(pbmc, ident.1 = 5, ident.2 = c(0, 3), min.pct = 0.25)
head(cluster5.markers, n = 5)
# find markers for every cluster compared to all remaining cells, report only the positive ones
pbmc.markers <- FindAllMarkers(pbmc, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
pbmc.markers %>%
    group_by(cluster) %>%
    slice_max(n = 2, order_by = avg_log2FC)
VlnPlot(pbmc, features = c("MS4A1", "CD79A"))
```
![](images/2021-10-25-15-46-35.png)
```r
# you can plot raw counts as well
VlnPlot(pbmc, features = c("NKG7", "PF4"), slot = "counts", log = TRUE)
```
![](images/2021-10-25-15-47-12.png)
```r
FeaturePlot(pbmc, features = c("MS4A1", "GNLY", "CD3E", "CD14", "FCER1A", "FCGR3A", "LYZ", "PPBP","CD8A"))
```
![](images/2021-10-25-15-47-32.png)
```r
# DoHeatmap()ä¸ºç»™å®šçš„ç»†èƒå’Œç‰¹å¾ç”Ÿæˆä¸€ä¸ªè¡¨è¾¾çƒ­å›¾ã€‚åœ¨è¿™ç§æƒ…å†µä¸‹ï¼Œæˆ‘ä»¬ä¸ºæ¯ä¸ªç°‡ç»˜åˆ¶å‰ 20 ä¸ªæ ‡è®°ï¼ˆæˆ–æ‰€æœ‰æ ‡è®°ï¼Œå¦‚æœå°äº 20ï¼‰ã€‚
pbmc.markers %>% group_by(cluster) %>%top_n(n = 10, wt = avg_log2FC) -> top10
DoHeatmap(pbmc, features = top10$gene) + NoLegend()
```
![](images/2021-10-25-15-47-46.png)

##### ä¸ä¼ªæ—¶é—´åˆ†æç›¸å…³çš„å·®å¼‚è¡¨è¾¾åˆ†æ 
> Monocle2åŒ…
1. å¯»æ‰¾éšä¼ªæ—¶é—´å˜åŒ–çš„åŸºå› 

Monocle çš„ä¸»è¦å·¥ä½œæ˜¯åœ¨ä¸çŸ¥é“æå‰æŸ¥çœ‹å“ªäº›åŸºå› çš„æƒ…å†µä¸‹ï¼Œå°†ç»†èƒæŒ‰ç”Ÿç‰©è¿‡ç¨‹ï¼ˆä¾‹å¦‚ç»†èƒåˆ†åŒ–ï¼‰çš„è¿›å±•é¡ºåºæ’åˆ—ã€‚å¯ä»¥åˆ†æç»†èƒä»¥æ‰¾åˆ°éšç€ç»†èƒè¿›å±•è€Œå‘ç”Ÿå˜åŒ–çš„åŸºå› ã€‚
```r
to_be_tested <- row.names(subset(fData(HSMM),
gene_short_name %in% c("MYH3", "MEF2C", "CCNB2", "TNNT1")))
cds_subset <- HSMM_myo[to_be_tested,]
# Monocle ä½¿ç”¨VGAMåŒ…å°†åŸºå› çš„è¡¨è¾¾æ°´å¹³å»ºæ¨¡ä¸ºä¼ªæ—¶é—´çš„å¹³æ»‘éçº¿æ€§å‡½æ•°
diff_test_res <- differentialGeneTest(cds_subset,
fullModelFormulaStr = "~sm.ns(Pseudotime)")
diff_test_res[,c("gene_short_name", "pval", "qval")]
plot_genes_in_pseudotime(cds_subset, color_by = "Hours")
```
![](images/2021-10-25-16-05-17.png)

2. åˆ†æå•ç»†èƒè½¨è¿¹ä¸­çš„åˆ†æ”¯
<br>å•ç»†èƒè½¨è¿¹åŒ…æ‹¬åˆ†æ”¯ã€‚åˆ†æ”¯çš„å‡ºç°æ˜¯å› ä¸ºç»†èƒæ‰§è¡Œæ›¿ä»£åŸºå› è¡¨è¾¾ç¨‹åºã€‚åˆ†æ”¯å‡ºç°åœ¨å‘è‚²è¿‡ç¨‹ä¸­çš„è½¨è¿¹ä¸­ï¼Œå½“ç»†èƒåšå‡ºå‘½è¿é€‰æ‹©æ—¶ï¼šä¸€ä¸ªå‘è‚²è°±ç³»æ²¿ç€ä¸€æ¡è·¯å¾„å‰è¿›ï¼Œè€Œå¦ä¸€ä¸ªè°±ç³»äº§ç”Ÿç¬¬äºŒæ¡è·¯å¾„ã€‚Monocleé‡‡ç”¨åˆ†æ”¯è¡¨è¾¾å¼åˆ†æå»ºæ¨¡ï¼Œä¸»è¦æ˜¯BEAM(Branched expression analysis modeling)å‡½æ•°ï¼Œå¯ä»¥å°†åˆ†å‰è¿‡ç¨‹é‡æ„ä¸ºä¸€ä¸ªåˆ†æ”¯è½¨è¿¹ï¼Œæ¯”è¾ƒåˆ†æç‚¹ä¸åˆ†ææœ«ç«¯çš„å·®å¼‚ï¼Œä»è€Œåˆ†æä¸åŒç»†èƒå‘½è¿ä¸‹çš„å·®å¼‚ã€‚<mark>Monocleå¯ä»¥æ¨¡æ‹Ÿå‡ºæ¯ä¸ªç»†èƒæ‰€å¤„çš„åˆ†åŒ–æ—¶é—´ï¼Œå¹¶å¯»æ‰¾éšç€åˆ†åŒ–æ—¶é—´é€æ¸å‡é«˜æˆ–é™ä½çš„åŸºå› è¡¨è¾¾çƒ­å›¾å’Œåˆ†å¸ƒå›¾ã€‚</mark>
```r
# BEAMå°†å·²æ’åºçš„CellDataSetorderCellså’Œè½¨è¿¹ä¸­åˆ†æ”¯ç‚¹çš„åç§°ä½œä¸ºè¾“å…¥ã€‚å®ƒè¿”å›æ¯ä¸ªåŸºå› çš„æ˜¾ç€æ€§åˆ†æ•°è¡¨ã€‚å¾—åˆ†æ˜¾ç€çš„åŸºå› åœ¨å…¶è¡¨è¾¾ä¸­è¢«è®¤ä¸ºæ˜¯åˆ†æ”¯ä¾èµ–æ€§çš„ã€‚
BEAM_res <- BEAM(lung, branch_point = 1, cores = 1)
BEAM_res <- BEAM_res[order(BEAM_res$qval),]
BEAM_res <- BEAM_res[,c("gene_short_name", "pval", "qval")]
# å¯ä»¥ä½¿ç”¨ç‰¹æ®Šç±»å‹çš„çƒ­å›¾å¯è§†åŒ–æ‰€æœ‰æ˜¾ç€ä¾èµ–äºåˆ†æ”¯çš„åŸºå› çš„å˜åŒ–ã€‚æ­¤çƒ­å›¾åŒæ—¶æ˜¾ç¤ºäº†ä¸¤ä¸ªè°±ç³»çš„å˜åŒ–ã€‚å®ƒè¿˜è¦æ±‚é€‰æ‹©è¦æ£€æŸ¥çš„åˆ†æ”¯ç‚¹ã€‚åˆ—æ˜¯ä¼ªæ—¶é—´ä¸­çš„ç‚¹ï¼Œè¡Œæ˜¯åŸºå› ï¼Œä¼ªæ—¶é—´çš„å¼€å§‹ä½äºçƒ­å›¾çš„ä¸­é—´ã€‚branch_point=1ï¼Œåˆ†æ”¯ç‚¹é€‰ä¸º1ï¼Œåˆ†æbranch_point = 1è¿™ä¸ªåˆ†æ”¯å¤„çš„ç»†èƒå‘½ååˆ†å‰æ˜¯å¦‚ä½•è¿›è¡Œçš„ã€‚num_clusters=4ï¼Œå°†åŸºå› æ ¹æ®è¡¨è¾¾ç›¸ä¼¼æ€§åˆ†æˆ4ä¸ªæ¨¡å—ã€‚

plot_genes_branched_heatmap(lung[row.names(subset(BEAM_res,
                                          qval < 1e-4)),],
                                          branch_point = 1,
                                          num_clusters = 4,
                                          cores = 1,
                                          use_gene_short_name = T,
                                          show_rownames = T)
                                          # å¯è§†åŒ–æ‰€æƒ³è¦ç ”ç©¶çš„åŸºå› 
lung_genes <- row.names(subset(fData(lung),
          gene_short_name %in% c("Ccnd2", "Sftpb", "Pdpn")))
plot_genes_branched_pseudotime(lung[lung_genes,],
                       branch_point = 1,
                       color_by = "Time",
                       ncol = 1)
```


> Monocle3åŒ…
1. <b>å›å½’åˆ†æï¼š
 <br></b>ä½¿ç”¨`fit_models()`ï¼Œå¯ä»¥è¯„ä¼°æ¯ä¸ªåŸºå› æ˜¯å¦ä¾èµ–äºæ—¶é—´ã€æ²»ç–—ç­‰å˜é‡ã€‚

```r
# é€‚ç”¨äºç ”ç©¶çš„ç”Ÿç‰©è¿‡ç¨‹ä¸­å‘ç°åŠ¨æ€è°ƒèŠ‚çš„åŸºå› 
gene_fits <- fit_models(cds_subset, model_formula_str = "~embryo.time")
fit_coefs <- coefficient_table(gene_fits)
# æ‰¾å‡ºå…·æœ‰é‡è¦æ—¶é—´åˆ†é‡çš„åŸºå› 
emb_time_terms %>% filter (q_value < 0.05) %>%
         select(gene_short_name, term, q_value, estimate)
# å¯è§†åŒ–ä¸Šè¿°æµ‹è¯•æ‰€æ­ç¤ºçš„å·®å¼‚
plot_genes_violin(cds_subset, group_cells_by="embryo.time.bin", ncol=2) +
      theme(axis.text.x=element_text(angle=45, hjust=1))
# æ§åˆ¶æ‰¹æ¬¡æ•ˆåº”å’Œå…¶ä»–å› ç´ 
gene_fits <- fit_models(cds_subset, model_formula_str = "~embryo.time + batch")
fit_coefs <- coefficient_table(gene_fits)
fit_coefs %>% filter(term != "(Intercept)") %>%
      select(gene_short_name, term, q_value, estimate)
# è¯„ä¼°åŸºå› è¡¨è¾¾æ¨¡å‹
evaluate_fits(gene_fits)
# å®Œæ•´æ¨¡å‹ï¼šçŸ¥é“æ¯ä¸ªç»†èƒè¢«æ”¶é›†çš„æ—¶é—´å’Œæ¥è‡ªå“ªä¸€æ‰¹
time_batch_models <- fit_models(cds_subset,
                                model_formula_str = "~embryo.time + batch",
                                expression_family="negbinomial")
# ç¼©å‡æ¨¡å‹ï¼šåªçŸ¥é“æ¯ä¸ªç»†èƒè¢«æ”¶é›†çš„æ—¶é—´
time_models <- fit_models(cds_subset,
                          model_formula_str = "~embryo.time",
                          expression_family="negbinomial")
compare_models(time_batch_models, time_models) %>% select(gene_short_name, q_value)
```
Monocleçš„fit_models()æ”¯æŒè´ŸäºŒé¡¹åˆ†å¸ƒå’Œä¸‹è¡¨ä¸­åˆ—å‡ºçš„å…¶ä»–å‡ ä¸ªåˆ†å¸ƒã€‚
<table>
   <tr>
      <td>quasipoisson</td>
      <td>å‡†æ³Šæ¾</td>
      <td>é»˜è®¤ä¸ºfit_models()</td>
   </tr>
   <tr>
      <td>negbinomial</td>
      <td>è´ŸäºŒé¡¹å¼</td>
      <td>æ¨èç”¨äºæ•°æ®é›†è¾ƒå°ï¼ˆå°‘äº 1,000 ä¸ªç»†èƒï¼‰</td>
   </tr>
   <tr>
      <td>poisson</td>
      <td>æ³Šæ¾</td>
      <td>ä¸å»ºè®®ã€‚ä»…ç”¨äºè°ƒè¯•å’Œæµ‹è¯•ã€‚</td>
   </tr>
   <tr>
      <td>binomial</td>
      <td>äºŒé¡¹å¼</td>
      <td>æ¨èç”¨äºå•ç»†èƒ ATAC-seq</td>
   </tr>
</table>

2. <b>å›¾è‡ªç›¸å…³åˆ†æï¼š<b>
<br>ä½¿ç”¨`graph_test()`ï¼Œå¯ä»¥æ‰¾åˆ°åœ¨è½¨è¿¹ä¸Šæˆ–ç°‡ä¹‹é—´å˜åŒ–çš„åŸºå› ã€‚

```r
##å¯»æ‰¾æ‹Ÿæ—¶è½¨è¿¹å·®å¼‚åŸºå› 
#graph_teståˆ†ææœ€é‡è¦çš„ç»“æœæ˜¯è«å…°æŒ‡æ•°ï¼ˆmorans_Iï¼‰ï¼Œå…¶å€¼åœ¨-1è‡³1ä¹‹é—´ï¼Œ0ä»£è¡¨æ­¤åŸºå› æ²¡æœ‰
#ç©ºé—´å…±è¡¨è¾¾æ•ˆåº”ï¼Œ1ä»£è¡¨æ­¤åŸºå› åœ¨ç©ºé—´è·ç¦»ç›¸è¿‘çš„ç»†èƒä¸­è¡¨è¾¾å€¼é«˜åº¦ç›¸ä¼¼ã€‚
Track_genes <- graph_test(cds, neighbor_graph="principal_graph", cores=10)
#æŒ‘é€‰top10ç”»å›¾å±•ç¤º
Track_genes_sig <- Track_genes %>% top_n(n=10, morans_I) %>%
                   pull(gene_short_name) %>% as.character()
#åŸºå› è¡¨è¾¾è¶‹åŠ¿å›¾
plot_genes_in_pseudotime(cds[Track_genes_sig,], color_cells_by="predicted.id", 
                              min_expr=0.5, ncol = 2)
#FeaturePlotå›¾
plot_cells(cds, genes=Track_genes_sig, show_trajectory_graph=FALSE,
               label_cell_groups=FALSE,  label_leaves=FALSE)
##å¯»æ‰¾å…±è¡¨è¾¾æ¨¡å—
genelist <- pull(Track_genes, gene_short_name) %>% as.character()
gene_module <- find_gene_modules(cds[genelist,], resolution=1e2, cores = 10)
cell_group <- tibble::tibble(cell=row.names(colData(cds)), 
                             cell_group=colData(cds)$predicted.id)
agg_mat <- aggregate_gene_expression(cds, gene_module, cell_group)
row.names(agg_mat) <- stringr::str_c("Module ", row.names(agg_mat))
pheatmap::pheatmap(agg_mat, scale="column", clustering_method="ward.D2")
```


#### åŸºå› é›†åˆ†æ
##### è¡¨è§‚é—ä¼ åˆ†æ
##### å¯Œé›†åˆ†æ
#### åŸºå› è°ƒæ§ç½‘ç»œåˆ†æ
##### åŸºå› å…±è¡¨è¾¾
##### è›‹ç™½è´¨äº’ä½œç½‘ç»œ
##### å›å½’æ¨¡å‹

###  
